// eslint-config-bragi/index.js
const a11y = require('./a11y');

/**
 * Expose estlint config
 */
module.exports = {
  extends: [
    'plugin:jsx-a11y/recommended',
    'plugin:nordic/recommended',
    'plugin:security/recommended',
    'airbnb',
  ],
  plugins: [
    'nordic',
    'react-hooks',
    'security',
  ],
  env: {
    // airbnb config already has { node: true, es6: true }, so just allow to use the `browser` stuff
    browser: true,
  },
  rules: {
    'max-len': ['warn', 120, 2, {
      ignoreUrls: true,
      ignoreComments: false,
    }],
    'no-param-reassign': ['error', { props: false }],
    'class-methods-use-this': 'off',
    'react/forbid-foreign-prop-types': ['warn'],
    'react/no-unused-prop-types': ['warn'],
    'react/prop-types': ['warn'],
    'react/jsx-filename-extension': 'off',
    'react-hooks/exhaustive-deps': ['warn'],
    'no-underscore-dangle': ['error', { allowAfterThis: true, allow: ['__PRELOADED_STATE__'] }],
    'arrow-parens': 'off',
    'import/prefer-default-export': 'off',
    'import/order': ['warn'],
    'object-curly-newline': 'off',
    'react/button-has-type': 'off',
    'react/jsx-one-expression-per-line': 'off',
    'react/jsx-props-no-spreading': 'off',
    'react/jsx-closing-tag-location': ['warn'],
    'react/jsx-wrap-multilines': ['error', { condition: 'parens', prop: 'parens' }],
    'react/require-default-props': 'off',
    'react/default-props-match-prop-types': 'off',
    'no-multiple-empty-lines': ['error', { max: 2, maxEOF: 1 }],
    'no-redeclare': ['error', { builtinGlobals: false }],
    ...a11y,
  },
  settings: {
    'import/core-modules': [
      'core-js',
      'newrelic',
      'prop-types',
      'react',
      'react-declarative-head',
      'react-dom',
      'react-side-effect',
      'serialize-javascript',
    ],
  },
  parser: '@babel/eslint-parser',
  parserOptions: {
    requireConfigFile: false,
  },
};
