// ml-web-navigation/src/navigation/header/props.js
const logoProps = require('./components/logo/props');
const meliLabProps = require('./components/meli-lab/props');
const userMenuProps = require('./components/user-menu/props');
const searchboxProps = require('./components/searchbox/props');
const cartProps = require('./components/cart/props');
const cpProps = require('./components/cp/props');
const navMenuProps = require('./components/nav-menu/props');
const notificationsProps = require('./components/notifications/props');
const exhibitorProps = require('./components/exhibitor/props');
const skipToMainContentProps = require('./components/skip-to-main-content/props');
const isMobile = require('../utils/is-mobile');
const isDesktop = require('../utils/is-desktop');
const downloadApp = require('../widgets/download-app/props');

const headerProps = (props, linkTagsProps) => {
  const { platform, type, cdn, internal } = props;

  return {
    type,
    internal,
    siteId: platform.siteId,
    isMobile: isMobile(props),
    isDesktop: isDesktop(props),
    cdnLink: cdn.link,
    logo: logoProps(props),
    skipToMainContent: skipToMainContentProps(props),
    melilab: meliLabProps(props),
    userMenu: userMenuProps(props, linkTagsProps),
    searchbox: searchboxProps(props),
    cart: cartProps(props),
    /**
     * With this `forceMobile: true` we are specifically saying that we always want to render
     * the CP component for web mobile, no matter if we are in desktop or web/mobile. This is due
     * to the fact that we want to be responsive and show the CP component when a user makes his/her
     * screen smaller in desktop.
     */
    cp: cpProps({ ...props, forceMobile: true }),
    exhibitor: exhibitorProps(props),
    notifications: notificationsProps(props),
    navMenu: navMenuProps(props, linkTagsProps),
    downloadApp: downloadApp(props),
  };
};

module.exports = headerProps;
