// first-party-web-navigation/src/navigation/css/index.js
const React = require('react');
const propTypes = require('prop-types');
const CommonsCss = require('@frontend-navigations/commons/lib/css-component');
const { DESKTOP } = require('@frontend-navigations/commons/constants/device-types');
const {
  DESKTOP: DESKTOP_CSS_FILE,
  MOBILE: MOBILE_CSS_FILE,
} = require('../../constants/css-files');

const deviceTypes = {
  [DESKTOP]: [{
    name: DESKTOP_CSS_FILE,
    media: '(min-width: 1024px)',
  }, {
    name: MOBILE_CSS_FILE,
    media: '(max-width: 1023px)',
  }],
  default: [{
    name: MOBILE_CSS_FILE,
  }],
};

const Css = ({
  navigationType,
  deviceType,
  firstRequestInlineStyles,
  platformId,
  siteId,
  embedCss,
  feConfigEnvironment,
  cdn,
  configGetter,
  configEnvironment,
  addConfigUpdateHandler,
}) => {
  // eslint-disable-next-line security/detect-object-injection
  const files = (deviceTypes[deviceType] || deviceTypes.default)
    .map(style => ({
      ...style,
      firstRequestInlineStyles,
    }));

  return (
    <CommonsCss
      files={files}
      navigationType={navigationType}
      platformId={platformId}
      siteId={siteId}
      embedCss={embedCss}
      feConfigEnvironment={feConfigEnvironment}
      cdn={cdn}
      configGetter={configGetter}
      configEnvironment={configEnvironment}
      addConfigUpdateHandler={addConfigUpdateHandler}
      configKey="firstParty.uiNavigationVersion"
      appName="first-party-web-navigation"
    />
  );
};

Css.propTypes = {
  firstRequestInlineStyles: propTypes.bool,
  deviceType: propTypes.string,
  navigationType: propTypes.string,
  platformId: propTypes.string,
  siteId: propTypes.string,
  embedCss: propTypes.bool,
  feConfigEnvironment: propTypes.string,
  cdn: propTypes.shape({
    link: propTypes.string,
  }),
  configGetter: propTypes.func,
  configEnvironment: propTypes.string,
  addConfigUpdateHandler: propTypes.func,
};

Css.defaultProps = {
  firstRequestInlineStyles: false,
  deviceType: null,
  navigationType: 'full',
  platformId: 'ML',
  siteId: 'MLA',
  embedCss: false,
  feConfigEnvironment: null,
  cdn: null,
  configGetter: null,
  configEnvironment: null,
  addConfigUpdateHandler: null,
};

module.exports = Css;
