// ml-web-navigation/src/prepare-to-render/services/index.js
const exhibitor = require('./exhibitor');
const cart = require('./cart');
const userMenu = require('./user-menu');
const notifications = require('./notifications');
const cp = require('./cp');
const loans = require('./loans');
const downloadApp = require('./download-app');
const pii = require('./pii');
const kyc = require('./kyc');
const req = require('./req');
const navigationFeatures = require('./navigation-features');
const type = require('./type');
const urls = require('./urls');

const services = [
  { key: 'exhibitorInfo', execute: exhibitor },
  { key: 'cartInfo', execute: cart },
  { key: 'userInfo', execute: userMenu },
  { key: 'notificationsInfo', execute: notifications },
  { key: 'cpInfo', execute: cp },
  { key: 'loansInfo', execute: loans },
  { key: 'downloadApp', execute: downloadApp },
  { key: 'navigationPIIData', execute: pii },
  { key: 'kycInfo', execute: kyc },
  { key: 'req', execute: req },
  { key: 'navigationUrls', execute: urls },
  { key: 'navigationFeatures', execute: navigationFeatures },
  { key: 'type', execute: type },
];

module.exports = services;
