// webpack-conformance-plugin/lib/rules/minification.js
/* eslint-disable class-methods-use-this */
const {
  CONFORMANCE_PREFIX,
  CONFORMANCE_RULE_FAILED,
  EARLY_EXIT_SUCCESS_RESULT,
} = require('../constants');

const { getLocalFileName } = require('../utils');

class WebpackConformanceRuleMinification {
  constructor(webpackConfig) {
    this.webpackConfig = webpackConfig;
  }

  buildStared(options) {
    const { optimization } = options;
    if (
      options.mode === 'production'
      && optimization
      && (optimization.minimize !== true
        || (optimization.minimizer && optimization.minimizer.length === 0))
    ) {
      const message = 'Minification is disabled for this build,'
        + ' disabling minification can result in serious performance degradation.';
      return {
        result: CONFORMANCE_RULE_FAILED,
        warnings: [
          {
            message: `${CONFORMANCE_PREFIX} minification: ${message}`,
            metadata: {
              column: 0,
              file: getLocalFileName(this.webpackConfig),
              line: 1,
              message,
              ruleId: 'webpack-conformance-rule-minification',
            },
          },
        ],
      };
    }
    return EARLY_EXIT_SUCCESS_RESULT;
  }
}

module.exports = WebpackConformanceRuleMinification;
