// mp-web-navigation/src/navigation/header/components/user-menu/navigation/index.js
/* eslint-disable jsx-a11y/no-redundant-roles */
const React = require('react');
const PropTypes = require('prop-types');
const Script = require('frontend-script');
const { DESKTOP, MOBILE } = require('@frontend-navigations/commons/constants/device-types');
const Link = require('../../../../utils/link');
const UserInfo = require('./user-info');
const LandingNavigationMobile = require('./landing-navigation/mobile');

const namespace = 'nav-header-user-menu';

const Navigation = (props) => {
  const {
    links,
    isLoggedIn,
    userInfo,
    deviceType,
    uiNavigationVersion,
    landingNavigation,
  } = props;

  return (
    <>
      <nav
        className={`${namespace} ${namespace}${isLoggedIn ? '--user' : '--guest'}`}
        id={`${namespace}-nav`}
      >
        {isLoggedIn && (<UserInfo {...userInfo} />)}
        {deviceType !== DESKTOP && <LandingNavigationMobile {...landingNavigation} />}
        {
          links.map(item => (
            <Link
              {...item}
              key={item.key}
              className={`${namespace}__link`}
            />
          ))
        }
        <span className={`${namespace}__top-arrow`} />
      </nav>
      { deviceType !== MOBILE && <Script src={`https://http2.mlstatic.com/frontend-assets/frontend-navigations/ui-navigation/${uiNavigationVersion}/mercadopago/user-menu.js`} /> }
    </>
  );
};

Navigation.propTypes = {
  links: PropTypes.arrayOf(
    PropTypes.shape({}),
  ).isRequired,
  userInfo: PropTypes.shape({}),
  isLoggedIn: PropTypes.bool,
  deviceType: PropTypes.string,
  uiNavigationVersion: PropTypes.string.isRequired,
  landingNavigation: PropTypes.shape({}),
};

Navigation.defaultProps = {
  userInfo: {},
  isLoggedIn: false,
  deviceType: 'desktop',
};

module.exports = Navigation;
