// cbt-web-navigation/src/navigation/types.js
const PropTypes = require('prop-types');

const siteUserType = {
  user_id: PropTypes.number.isRequired,
  site_id: PropTypes.string.isRequired,
  status: PropTypes.string.isRequired,
  enabled_sites: PropTypes.arrayOf(PropTypes.string),
  parent_id: PropTypes.number.isRequired,
  logistic_type: PropTypes.string.isRequired,
};

const cbtInfoType = {
  siteUsers: PropTypes.arrayOf(PropTypes.shape(siteUserType)),
  siteUserId: PropTypes.number.isRequired,
  currentSiteId: PropTypes.string.isRequired,
};

const userType = {
  id: PropTypes.number.isRequired,
  nickname: PropTypes.string.isRequired,
  active: PropTypes.bool.isRequired,
};

const navigationPIIData = {
  userName: PropTypes.string.isRequired,
};

const linkType = {
  name: PropTypes.string,
  link: PropTypes.string,
  id: PropTypes.string,
};

const iconType = {
  icon: PropTypes.string.isRequired,
  name: PropTypes.string,
  link: PropTypes.string.isRequired,
  showText: PropTypes.bool,
  id: PropTypes.string.isRequired,
};

const siteType = {
  text: PropTypes.string.isRequired,
  countryId: PropTypes.string.isRequired,
  flag: PropTypes.string,
};

const logisticType = {
  remote: PropTypes.string,
  '3pl': PropTypes.string.isRequired,
  fulfillment: PropTypes.string.isRequired,
};

const userMenuType = {
  cbtInfo: PropTypes.shape(cbtInfoType),
  links: PropTypes.objectOf(PropTypes.shape(linkType)),
  sites: PropTypes.objectOf(PropTypes.shape(siteType)),
  logisticTypes: PropTypes.shape(logisticType),
  loggedLinks: PropTypes.arrayOf(PropTypes.shape(linkType)),
  icons: PropTypes.arrayOf(PropTypes.shape(iconType)),
};

module.exports = {
  cbtInfoType,
  userMenuType,
  siteUserType,
  userType,
  logisticType,
  siteType,
  linkType,
  iconType,
  navigationPIIData,
};
