// ml-web-navigation/src/navigation/header/components/user-menu/mobile/andes-money-amount/lib/constants/currencies.js
const DEFAULT_CURRENCY_SYMBOL = '$';
const DEFAULT_CURRENCY_SYMBOL_TEXT = { singular: 'peso', plural: 'pesos' };
const DEFAULT_CENTS_TEXT = { singular: 'centavo', plural: 'centavos' };

const CURRENCY_SYMBOLS_BY_ID = {
  ARS: '$',
  BOB: 'Bs',
  BRL: 'R$',
  CLF: 'UF',
  CLP: '$',
  COP: '$',
  CRC: '₡',
  CUC: 'CUC',
  CUP: '$',
  DOP: '$',
  EUR: '€',
  GTQ: 'Q',
  HNL: 'L',
  MXN: '$',
  NIO: 'C$',
  PAB: 'B/.',
  PEN: 'S/',
  PYG: '₲',
  USD: 'U$S',
  UYU: '$',
  VEF: 'Bs.',
  VES: 'Bs.',
};

const CURRENCY_SYMBOLS = ['$', 'Bs', 'R$', 'UF', '₡', 'CUC', '€', 'Q', 'L', 'C$', 'B/.', 'S/', '₲', 'U$S', 'Bs.'];

const CRYPTO_CURRENCY_IDS = ['BTC', 'ETH', 'USDP'];
const DEFAULT_CURRENCY_IDS = [
  'ARS',
  'BOB',
  'BRL',
  'CLF',
  'CLP',
  'COP',
  'CRC',
  'CUC',
  'CUP',
  'DOP',
  'EUR',
  'GTQ',
  'HNL',
  'MXN',
  'NIO',
  'PAB',
  'PEN',
  'PYG',
  'USD',
  'UYU',
  'VEF',
  'VES',
];

const CURRENCY_IDS = [...DEFAULT_CURRENCY_IDS, ...CRYPTO_CURRENCY_IDS];

const CURRENCY_TEXT_BY_SYMBOL = {
  $: DEFAULT_CURRENCY_SYMBOL_TEXT,
  Bs: {
    singular: 'boliviano',
    plural: 'bolivianos',
  },
  R$: {
    singular: 'real',
    plural: 'reales',
  },
  UF: {
    singular: 'unidad de fomento',
    plural: 'unidades de fomento',
  },
  '₡': {
    singular: 'colón',
    plural: 'colones',
  },
  CUC: DEFAULT_CURRENCY_SYMBOL_TEXT,
  '€': {
    singular: 'euro',
    plural: 'euros',
  },
  Q: {
    singular: 'quetzal guatemalteco',
    plural: 'quetzales guatemaltecos',
  },
  L: {
    singular: 'lempira',
    plural: 'lempiras',
  },
  C$: {
    singular: 'córdoba',
    plural: 'córdobas',
  },
  'B/.': {
    singular: 'balboa',
    plural: 'balboas',
  },
  'S/': {
    singular: 'sol',
    plural: 'soles',
  },
  '₲': {
    singular: 'guaraní',
    plural: 'guaraníes',
  },
  U$S: {
    singular: 'dólar',
    plural: 'dólares',
  },
  'Bs.': {
    singular: 'bolívar',
    plural: 'bolívares',
  },
};

const CURRENCY_TEXT_BY_ID = {
  ARS: DEFAULT_CURRENCY_SYMBOL_TEXT,
  BOB: {
    singular: 'boliviano',
    plural: 'bolivianos',
  },
  BRL: {
    singular: 'real',
    plural: 'reales',
  },
  BTC: 'bitcoins',
  CLF: {
    singular: 'unidad de fomento',
    plural: 'unidades de fomento',
  },
  CLP: DEFAULT_CURRENCY_SYMBOL_TEXT,
  COP: DEFAULT_CURRENCY_SYMBOL_TEXT,
  CRC: {
    singular: 'colón',
    plural: 'colones',
  },
  CUC: DEFAULT_CURRENCY_SYMBOL_TEXT,
  CUP: DEFAULT_CURRENCY_SYMBOL_TEXT,
  DOP: DEFAULT_CURRENCY_SYMBOL_TEXT,
  EUR: {
    singular: 'euro',
    plural: 'euros',
  },
  ETH: 'ethereum',
  GTQ: {
    singular: 'quetzal guatemalteco',
    plural: 'quetzales guatemaltecos',
  },
  HNL: {
    singular: 'lempira',
    plural: 'lempiras',
  },
  MXN: DEFAULT_CURRENCY_SYMBOL_TEXT,
  NIO: {
    singular: 'córdoba',
    plural: 'córdobas',
  },
  PAB: {
    singular: 'balboa',
    plural: 'balboas',
  },
  PEN: {
    singular: 'sol',
    plural: 'soles',
  },
  PYG: {
    singular: 'guaraní',
    plural: 'guaraníes',
  },
  USD: {
    singular: 'dólar',
    plural: 'dólares',
  },
  USDP: 'USDP Stablecoin',
  UYU: DEFAULT_CURRENCY_SYMBOL_TEXT,
  VEF: {
    singular: 'bolívar',
    plural: 'bolívares',
  },
  VES: {
    singular: 'bolívar',
    plural: 'bolívares',
  },
};

const CENTS_TEXT_BY_SYMBOL = {
  $: DEFAULT_CENTS_TEXT,
  Bs: DEFAULT_CENTS_TEXT,
  R$: DEFAULT_CENTS_TEXT,
  UF: DEFAULT_CENTS_TEXT,
  '₡': {
    singular: 'céntimo',
    plural: 'céntimos',
  },
  CUC: DEFAULT_CENTS_TEXT,
  '€': DEFAULT_CENTS_TEXT,
  Q: DEFAULT_CENTS_TEXT,
  L: DEFAULT_CENTS_TEXT,
  C$: DEFAULT_CENTS_TEXT,
  'B/.': {
    singular: 'centésimo',
    plural: 'centésimos',
  },
  'S/': {
    singular: 'céntimo',
    plural: 'céntimos',
  },
  '₲': {
    singular: 'céntimo',
    plural: 'céntimos',
  },
  U$S: DEFAULT_CENTS_TEXT,
  'Bs.': DEFAULT_CENTS_TEXT,
};

const CENTS_TEXT_BY_ID = {
  ARS: DEFAULT_CENTS_TEXT,
  BOB: DEFAULT_CENTS_TEXT,
  BRL: DEFAULT_CENTS_TEXT,
  CLF: DEFAULT_CENTS_TEXT,
  CLP: DEFAULT_CENTS_TEXT,
  COP: DEFAULT_CENTS_TEXT,
  CRC: {
    singular: 'céntimo',
    plural: 'céntimos',
  },
  CUC: DEFAULT_CENTS_TEXT,
  CUP: DEFAULT_CENTS_TEXT,
  DOP: DEFAULT_CENTS_TEXT,
  EUR: DEFAULT_CENTS_TEXT,
  GTQ: DEFAULT_CENTS_TEXT,
  HNL: DEFAULT_CENTS_TEXT,
  MXN: DEFAULT_CENTS_TEXT,
  NIO: DEFAULT_CENTS_TEXT,
  PAB: {
    singular: 'centésimo',
    plural: 'centésimos',
  },
  PEN: {
    singular: 'céntimo',
    plural: 'céntimos',
  },
  PYG: {
    singular: 'céntimo',
    plural: 'céntimos',
  },
  USD: DEFAULT_CENTS_TEXT,
  UYU: {
    singular: 'centésimo',
    plural: 'centésimos',
  },
  VEF: DEFAULT_CENTS_TEXT,
  VES: DEFAULT_CENTS_TEXT,
};

module.exports = {
  CURRENCY_SYMBOLS,
  CURRENCY_IDS,
  DEFAULT_CURRENCY_IDS,
  CURRENCY_SYMBOLS_BY_ID,
  CURRENCY_TEXT_BY_SYMBOL,
  CURRENCY_TEXT_BY_ID,
  CENTS_TEXT_BY_SYMBOL,
  CENTS_TEXT_BY_ID,
  DEFAULT_CURRENCY_SYMBOL,
  DEFAULT_CURRENCY_SYMBOL_TEXT,
  DEFAULT_CENTS_TEXT,
  CRYPTO_CURRENCY_IDS,
};
