// ml-web-navigation/src/navigation/header/components/user-menu/mobile/guest/welcome/index.js
const React = require('react');
const PropTypes = require('prop-types');
const Icon = require('./icon');

const namespace = 'nav-header-menu-mobile-guest';
const navMobileButton = 'nav-mobile-button';
const GuestWelcome = (props) => {
  const { login, registration, title, subtitle } = props;

  return (
    <div id="nav-header-menu-mobile-user-info" className={namespace}>
      <Icon namespace={namespace} />
      <h3 className={`${namespace}-title`}>{title}</h3>
      <p className={`${namespace}-text`}>{subtitle}</p>
      <div className={`${namespace}-buttons`}>
        <a rel="nofollow" className={`${navMobileButton} ${navMobileButton}-filled`} href={login.link}>{login.name}</a>
        <a rel="nofollow" className={`${navMobileButton} ${navMobileButton}-outline`} href={registration.link}>{registration.name}</a>
      </div>
    </div>
  );
};

GuestWelcome.propTypes = {
  login: PropTypes.shape({
    link: PropTypes.string.isRequired,
    name: PropTypes.string.isRequired,
  }).isRequired,
  registration: PropTypes.shape({
    link: PropTypes.string.isRequired,
    name: PropTypes.string.isRequired,
  }).isRequired,
  title: PropTypes.string.isRequired,
  subtitle: PropTypes.string.isRequired,
};

module.exports = GuestWelcome;
