// vis-web-navigation/src/navigation/header/components/user-menu/index.js
const React = require('react');
const PropTypes = require('prop-types');
const MobileUserMenu = require('./mobile');
const DesktopUserMenu = require('./desktop');

const UserMenu = (props) => {
  const {
    navMenuSwitchId,
    isUserLoggedIn,
    guestMenu,
    loggedInMenu,
    isDesktop,
  } = props;

  return (
    <>
      <MobileUserMenu
        isUserLoggedIn={isUserLoggedIn}
        navMenuSwitchId={navMenuSwitchId}
        guestMenu={guestMenu && guestMenu.mobile}
        loggedInMenu={loggedInMenu && loggedInMenu.mobile}
      />
      { isDesktop && (
        <DesktopUserMenu
          isUserLoggedIn={isUserLoggedIn}
          navMenuSwitchId={navMenuSwitchId}
          guestMenu={guestMenu && guestMenu.desktop}
          loggedInMenu={loggedInMenu && loggedInMenu.desktop}
        />)}
    </>
  );
};

UserMenu.propTypes = {
  navMenuSwitchId: PropTypes.string.isRequired,
  isUserLoggedIn: PropTypes.bool.isRequired,
  isDesktop: PropTypes.bool.isRequired,
  guestMenu: PropTypes.shape({
    mobile: PropTypes.object.isRequired,
  }),
  loggedInMenu: PropTypes.shape({
    mobile: PropTypes.object.isRequired,
  }),
};

UserMenu.defaultProps = {
  guestMenu: null,
  loggedInMenu: null,
  isDesktop: false,
};

module.exports = UserMenu;
