// perfill-agent/src/constants/metrics.js
module.exports = {
  AVAILABILITY_FAILURE: 'failure',
  AVAILABILITY_TRANSFER_SIZE: 'transfersize',
  WINDOW_DOM: 'window-dom',
  D2ID: 'd2id',
  NETWORK_TYPE: 'network-type',
  NETWORK_SPEED: 'network-speed',
  USER_AGENT: 'user-agent',
  RAM_SIZE: 'ram-size',
  WINDOW_WIDTH: 'window-width',
  WINDOW_HEIGHT: 'window-height',
  LCP: 'lcp',
  CLS: 'cls',
  FCP: 'fcp',
  FID: 'fid',
  TBT: 'tbt',
  TBT_FINAL: 'tbtFinal',
  LCP_FINAL: 'lcpFinal',
  CLS_FINAL: 'clsFinal',
  TTFB: 'time-to-first-byte',
  DNS_LOOKUP_TIME: 'dns-lookup-time',
  DOWNLOAD_TIME: 'download-time',
  FETCH_TIME: 'fetch-time',
  HEADER_SIZE: 'header-size',
  TOTAL_TIME: 'total-time',
  WORKER_TIME: 'worker-time',
  HYDRATE_COMPLETE: 'hydrate-complete',
  HYDRATE_DURATION: 'hydrate-duration',
  REDIRECT_COUNT: 'redirect-count',
  DOM_COMPLETE: 'dom-complete',
  DOM_INTERACTIVE: 'dom-interactive',
  DOM_CONTENT_LOADED_END: 'dom-content-loaded-end',
  DOM_CONTENT_LOADED_START: 'dom-content-loaded-start',
  DOM_CONTENT_LOADED: 'dom-content-loaded',
  ONLOAD: 'onload',
  ONLOAD_END: 'onload-end',
  ONLOAD_START: 'onload-start',
  REQUEST_START: 'request-start',
  RESPONSE_START: 'response-start',
  RESPONSE_END: 'response-end',
  RESPONSE_TIME: 'response-time',
  NAVIGATION_TYPE: 'navigation-type',
  INITIAL_CONTENT_SIZE: 'initial-content-size',
  DARK_MODE: 'dark-mode',
  REDUCED_MOTION: 'reduced-motion',
  REDUCED_COLOR: 'reduced-contrast',
  URL: 'url',
  X_CDN: 'cdn',
  DURATION_SUM: 'duration-sum',
  RESOURCE_SIZE: 'resources-size',
  RESOURCE_FAILURE: 'resources-failure',
  RESOURCE_SUCCESS: 'resources-success',
};
