// mp-web-navigation/src/navigation/footer/components/navigation/index.js
const React = require('react');
const PropTypes = require('prop-types');
const classNames = require('classnames');
const Link = require('../../../utils/link');
const ReclamationBook = require('../reclamationBook');
const tracking = require('../../../../constants/tracking');

const NavSection = ({
  id,
  className,
  title,
  links,
  siteId,
  reclamationBook,
}) => (
  <section
    className={classNames(
      'nav-section-mp',
      { [`nav-section-mp--${id.replace(/_/g, '-')}`]: id },
      { [`nav-section-mp--${className}`]: className },
    )}
  >
    {title && title?.href ? (
      <a
        className="nav-section-mp__title nav-section-mp__title-link"
        data-track-action={tracking.ACTION}
        data-track-id={id.replace(/_/g, '-')}
        href={title?.href}
      >
        {title?.text}
      </a>
    ) : (
      <h2 className="nav-section-mp__title">{title?.text}</h2>
    )}
    <nav className="nav-section-mp__nav">
      {links.map((item) => (
        <Link
          {...item}
          rel="noopener noreferrer"
          link={item.link.href}
          key={item.id}
          className="nav-section-mp__nav-link"
        />
      ))}
    </nav>
    {siteId === 'MPE' && Object.entries(reclamationBook).length && (
      <ReclamationBook {...reclamationBook} />
    )}
  </section>
);

NavSection.propTypes = {
  id: PropTypes.string,
  className: PropTypes.string,
  title: PropTypes.shape({
    text: PropTypes.string,
    href: PropTypes.string,
  }),
  links: PropTypes.arrayOf(
    PropTypes.shape({
      name: PropTypes.string.isRequired,
      link: PropTypes.shape({
        href: PropTypes.string.isRequired,
      }).isRequired,
      icon: PropTypes.string,
      id: PropTypes.string.isRequired,
      selected: PropTypes.bool,
    }),
  ),
  siteId: PropTypes.string,
  reclamationBook: PropTypes.shape({}),
};

NavSection.defaultProps = {
  id: '',
  className: '',
  title: null,
  links: [],
  siteId: '',
  reclamationBook: null,
};

module.exports = NavSection;
