// vis-web-navigation/src/navigation/header/components/base-header/index.js
/* eslint-disable react/forbid-prop-types */
const React = require('react');
const PropTypes = require('prop-types');
const classNames = require('classnames');
const Logo = require('../logo');

const namespace = 'nav-header';
const navBounds = 'nav-bounds';
const BaseHeader = (props) => {
  const {
    children,
    siteId,
    type,
    showLogo,
    logo,
    melilab,
  } = props;
  const className = classNames(namespace, `${namespace}-${type}`);

  return (
    <header
      role="banner"
      data-siteid={siteId}
      className={className}
    >
      <div className={navBounds}>
        { showLogo && <Logo url={logo.url} label={logo.label} melilab={melilab} /> }
        {children}
      </div>
    </header>
  );
};

BaseHeader.propTypes = {
  children: PropTypes.oneOfType([
    PropTypes.arrayOf(PropTypes.node),
    PropTypes.node,
  ]),
  logo: PropTypes.shape({
    url: PropTypes.string.isRequired,
    label: PropTypes.string.isRequired,
  }).isRequired,
  siteId: PropTypes.string.isRequired,
  type: PropTypes.string.isRequired,
  showLogo: PropTypes.bool,
  melilab: PropTypes.object.isRequired,
};


BaseHeader.defaultProps = {
  children: null,
  showLogo: false,
  showUserMenu: false,
};

module.exports = BaseHeader;
