// cbt-web-navigation/src/navigation/css/index.js
const React = require('react');
const propTypes = require('prop-types');
const CommonsCss = require('@frontend-navigations/commons/lib/css-component');
const {
  FLOW_LITE,
  FLOW,
  FULL,
  LITE,
  SIMPLE,
} = require('../../constants/navigation-types');
const {
  NAVIGATION_FLOW_LITE,
  NAVIGATION_FLOW,
  NAVIGATION_FULL,
  NAVIGATION_LITE,
  NAVIGATION_SIMPLE,
} = require('../../constants/css-files');

const styleSheetsHash = {
  [FLOW_LITE]: NAVIGATION_FLOW_LITE,
  [FLOW]: NAVIGATION_FLOW,
  [FULL]: NAVIGATION_FULL,
  [LITE]: NAVIGATION_LITE,
  [SIMPLE]: NAVIGATION_SIMPLE,
};

const Css = ({
  navigationType,
  firstRequestInlineStyles,
  platformId,
  siteId,
  embedCss,
  feConfigEnvironment,
  cdn,
  configGetter,
  configEnvironment,
  addConfigUpdateHandler,
}) => {
  const name = styleSheetsHash[navigationType]; // eslint-disable-line security/detect-object-injection
  const files = !name
    ? []
    : [{
      name,
      firstRequestInlineStyles,
    }];

  return (
    <CommonsCss
      files={files}
      navigationType={navigationType}
      platformId={platformId}
      siteId={siteId}
      embedCss={embedCss}
      feConfigEnvironment={feConfigEnvironment}
      cdn={cdn}
      configGetter={configGetter}
      configEnvironment={configEnvironment}
      addConfigUpdateHandler={addConfigUpdateHandler}
      appName="cbt-web-navigation"
    />
  );
};

Css.propTypes = {
  firstRequestInlineStyles: propTypes.bool,
  navigationType: propTypes.string,
  platformId: propTypes.string,
  siteId: propTypes.string,
  embedCss: propTypes.bool,
  feConfigEnvironment: propTypes.string,
  cdn: propTypes.shape({
    link: propTypes.string,
  }),
  configGetter: propTypes.func,
  configEnvironment: propTypes.string,
  addConfigUpdateHandler: propTypes.func,
};

Css.defaultProps = {
  firstRequestInlineStyles: false,
  navigationType: FULL,
  platformId: 'ML',
  siteId: 'CBT',
  embedCss: false,
  feConfigEnvironment: null,
  cdn: null,
  configGetter: null,
  configEnvironment: null,
  addConfigUpdateHandler: null,
};

module.exports = Css;
