// frontend-config/utils/migration/configurations/ml.js
const config = require('ml-navigation-config');
const { linkMapper } = require('./mappers/link');
const { identityMapper } = require('./mappers/identity');
const { footerAccessMapper } = require('./mappers/footer-access');

const concatLinks = links => (links || []).reduce((tempLinks, linksList) => tempLinks.concat(linksList), []);
const mobileHeaderMenuLinks = (value) => concatLinks((value || {}).logged)
  .concat(concatLinks((value || {}).guest))
  .map(link => `url.${link}`);

module.exports = {
  config,
  rules: [{
    platforms: [{
      id: 'ML',
      excludedSites: [
        'CBT',
      ],
    }],
    keys: [{
      key: 'uiNavigationVersion'
    }, {
      key: 'siteSpecificCSS'
    }, {
      key: 'linkTags'
    }, {
      key: 'sic'
    }, {
      key: 'kyc'
    }, {
      key: 'skipToMainContent'
    }, {
      key: 'mobileHeaderMenuIcons'
    }, {
      key: 'url.changeNickName'
    }, {
      key: 'url.downloadAppLanding'
    }, {
      key: 'url.downloadAppStore'
    }, {
      key: 'url.myAccount'
    }, {
      key: 'url.honeyPath'
    }, {
      key: 'url.cart'
    }, {
      key: 'url.cpLocation'
    }, {
      key: 'url.purchases'
    }, {
      key: 'url.helpMain'
    }, {
      key: 'url.login'
    }, {
      key: 'url.logout'
    }, {
      key: 'url.registration'
    }, {
      key: 'url.bookmarks'
    }, {
      key: 'url.notificationsCenter'
    }, {
      key: 'url.loyalty'
    }, {
      key: 'url.credits'
    }, {
      key: 'url.listing'
    }, {
      key: 'url.companyFAQ'
    }, {
      key: 'userMenuGuest'
    }, {
      key: 'extraCompanyName'
    }, {
      key: 'companyName'
    }, {
      key: 'copyrightSince'
    }, {
      key: 'mobileAppsStore'
    }, {
      key: 'defaultTitle'
    }, {
      key: 'footerAccess',
      mapper: footerAccessMapper,
    }, {
      key: 'mobileFooterLinks',
      mapper: linkMapper,
    }, {
      key: 'mobileSecondaryFooterLinks',
      mapper: linkMapper,
    }, {
      key: 'guestMobileFooterLinks',
      mapper: linkMapper,
    }, {
      key: 'guestMobileFooterSecondaryLinks',
      mapper: linkMapper,
    }, {
      key: 'desktopFooterLinks',
      mapper: identityMapper,
    }, {
      key: 'mobileHeaderMenuLinks',
      mapper: mobileHeaderMenuLinks,
    }, {
      key: 'headerMenuLinks',
      mapper: linkMapper,
    }],
  }],
};
