// frontend-config/utils/migration/configurations/mp.js
const config = require('mp-navigation-config');
const { identityMapper } = require('./mappers/identity');
const { identityString } = require('./mappers/identity-string');
const { keyMapper } = require('./mappers/key');

module.exports = {
  config,
  rules: [{
    platforms: [{
      id: 'MP',
      excludedSites: [
        'CBT',
      ],
    }],
    keys: [{
      key: 'uiNavigationVersion',
    }, {
      key: "companyName",
    }, {
      key: "extraCompanyName",
    }, {
      key: "defaultTitle",
    },{
      key: 'siteSpecificCSS',
    }, {
      key: 'navHeader',
    }, {
      key: 'url.helpClaims',
    }, {
      key: 'url.about',
    }, {
      key: 'url.developers',
    }, {
      key: 'url.commercial',
    }, {
      key: 'performanceFooterLinks',
      mapper: identityMapper,
    }, {
      key: 'desktopFooterLinks',
      mapper: identityMapper,
    }, {
      key: 'footer.social',
      mapper: identityMapper,
    }, {
      key: 'navHeader.userInfo.redirect',
      mapper: identityString,
    }, {
      key: 'navHeader.menuLinks.logged',
      mapper: keyMapper,
    }, {
      key: 'navHeader.menuLinks.guest',
      mapper: keyMapper,
    }, {
      key: 'navHeader.itemsLinks.logged',
      mapper: keyMapper,
    }, {
      key: 'navHeader.itemsLinks.guest',
      mapper: keyMapper,
    }, {
      key: 'navHeader.mobileItemsLinks.logged',
      mapper: keyMapper,
    }, {
      key: 'navHeader.mobileItemsLinks.guest',
      mapper: keyMapper,
    }],
  }],
};
