// ml-web-navigation/src/navigation/header/components/user-menu/mobile/logged-in/melicoins/coins.js
const React = require('react');
const Image = require('frontend-image');
const PropTypes = require('prop-types');
const MoneyAmount = require('../../andes-money-amount');

const namespace = 'user-menu__coins';

const Coins = ({
  id,
  icon,
  label,
  exchange,
  units,
  variation,
  currency,
  symbol,
  cents,
  action,
  exchangeRate,
  index,
  array,
  kyc,
}) => {
  const kycLabel = kyc !== undefined ? kyc.label : '';
  const renderLineTop = index === 0 ? 'user-menu__coins-divisor' : '';
  const renderLineBot = index === array.length - 1 && kycLabel !== undefined && action === undefined
    ? 'user-menu__coins-divisor'
    : '';
  const renderLineCoins = index === array.length - 1 ? '' : 'user-menu__coins-divisor-coins';

  return (
    <li id={id} className={`${namespace}-item`}>
      <span className={`${renderLineTop}`} />

      <div className={`${namespace}-each-items`}>
        <Image
          src={icon}
          alt=""
          className={`${namespace}-icon ${namespace}-icon-mobile ${namespace}-title-image-align`}
          onload="off"
        />

        <div className={`${namespace}-item`}>
          <div className={`${namespace}-item-core`}>
            <div
              className={`${namespace}-item-container-title ${namespace}-item-container-title-mobile`}
            >
              <span
                className={`${namespace}-title  ${namespace}-title-weight `}
              >
                {label}
              </span>
            </div>

            <div className={`${namespace}-title ${namespace}-title-weight`}>
              <MoneyAmount
                className={`${namespace}-currency`}
                centsType="superscript"
                size={16}
                compact
                amount={{
                  symbol,
                  currencyId: currency,
                  fraction: units,
                  cents,
                }}
              />
            </div>
          </div>

          {exchange && (
            <div
              className={`${namespace}-container-item-info ${namespace}-item-core`}
            >
              <span className={`${namespace}-subtitle`} aria-hidden="true">
                {`${exchange.label} `}
              </span>
              <span className="andes-visually-hidden">
                {' '}
                {exchange.accessibility_label}{' '}
              </span>
              <span
                className={`${namespace}-subtitle ${namespace}-subtitle-ma`}
              >
                <span>
                  {symbol} {exchangeRate}
                </span>
              </span>
              <span
                className={`${namespace}-percentage
                ${namespace}-percentage-mobile
                ${namespace}-percentage-${variation.color}`}
                aria-hidden="true"
              >
                {variation.label}
              </span>
              <span className="andes-visually-hidden">
                {variation.accessibility_label}
              </span>
            </div>
          )}

          {action !== undefined && (
            <span id="user-menu__coins-mp" className="user-menu__coins-mp-link">
              <a
                href={action.target}
                className={`${namespace}-kyc-label ${namespace}-kyc-label-mobile ${namespace}-info-link`}
              >
                <span>{action.label}</span>
              </a>
            </span>
          )}
        </div>
      </div>
      <span className={`${renderLineCoins} ${renderLineBot}`} />
    </li>
  );
};

Coins.propTypes = {
  id: PropTypes.string,
  icon: PropTypes.string,
  label: PropTypes.string,
  exchange: PropTypes.object,
  units: PropTypes.string,
  variation: PropTypes.string,
  currency: PropTypes.string,
  symbol: PropTypes.string,
  cents: PropTypes.string,
  action: PropTypes.string,
  exchangeRate: PropTypes.string,
  index: PropTypes.string,
  array: PropTypes.array,
  kyc: PropTypes.string,
};

module.exports = Coins;
