// ml-web-navigation/src/navigation/widgets/user-menu/index.js
/**
 * Module dependencies
 */
const React = require('react');
const PropTypes = require('prop-types');
const Script = require('frontend-script');
const tabIndex = require('../../utils/tab-index');

const UserMenuWidget = (props) => {
  const { show, domain, baseUrl, version, searchFocus, userInfo } = props;

  return show && (
    <Script src={`${baseUrl}/frontend-assets/nav-widget-ml-user-menu/v${version}/user-menu.js`}>{`
        (function(win, freya) {
          if (win.UserMenuWidget && typeof freya !== 'undefined') {
            new UserMenuWidget({
              mainDomain: '${domain}',
              bus: freya,
              tabIndex: ${tabIndex('user-menu-link', searchFocus)},
              data: ${JSON.stringify(userInfo)}
            });
          }
        })(window, window.freya);
    `}
    </Script>
  );
};

UserMenuWidget.propTypes = {
  domain: PropTypes.string.isRequired,
  show: PropTypes.bool,
  baseUrl: PropTypes.string.isRequired,
  version: PropTypes.string,
  searchFocus: PropTypes.bool,
  userInfo: PropTypes.shape({}),
};

UserMenuWidget.defaultProps = {
  version: '1.6.0',
  show: false,
  searchFocus: false,
  userInfo: null,
};

module.exports = UserMenuWidget;
