// vis-web-navigation/src/navigation/footer/props.js
/* eslint-disable max-len */
const feConfig = require('frontend-config');
const copyrightProps = require('./components/copyright/props');
const baseFooterProps = require('./components/base-footer/props');
const footerAccessProps = require('./components/footer-access/props');
const footerLinksProps = require('./components/footer-links/props');
const userMenuProps = require('./components/user-menu/props');
const footerNavigationProps = require('./components/footer-navigation/props');
const honeypathTrap = require('./components/honeypath-trap/props');
const platformNameByBrand = require('./components/brand/constants');

const footerProps = (props) => {
  const { platform, feConfigEnvironment } = props;
  const { id, siteId } = platform;
  let platformName = feConfig.get('platformName', id, siteId, feConfigEnvironment);
  const baseCopyrightProps = copyrightProps(props);
  const basefooterProps = baseFooterProps(props);
  const isDesktop = props.device.desktop || false;
  platformName = platformNameByBrand[id] || platformName;

  if (isDesktop) {
    baseCopyrightProps.companyName = `${baseCopyrightProps.companyName} ${basefooterProps.extraCompanyName || ''}`;
    basefooterProps.extraCompanyName = null;
  }

  return {
    isDesktop,
    type: props.type,
    copyright: baseCopyrightProps,
    baseFooter: basefooterProps,
    userMenu: userMenuProps(props),
    footerAccess: footerAccessProps(props),
    footerNavigation: footerNavigationProps(props),
    footerLinks: footerLinksProps(props),
    honeypathTrap: honeypathTrap(props),
    platformName,
  };
};

module.exports = footerProps;
