// frontend-melidata/src/services/melidata-record.js
const config = require('frontend-config');
const log = require('frontend-logger')('melidata');
const restclient = require('frontend-restclient');
const buildFlowStarterContext = require('frontend-restclient/src/build-flow-context');

const {
  DEFAULT_MELIDATA_ML_ENDPOINT,
  DO_POST,
} = require('../../config/melidata');

class MelidataRecordService {
  constructor() {
    this.doPost = DO_POST;
    this.url = config.get('melidata.endpoint') || DEFAULT_MELIDATA_ML_ENDPOINT;
  }

  postToMelidata(track) {
    const trackToSend = track.toMap();

    if (this.doPost) {
      log.debug(`Posting to melidata api. Tracks : ${trackToSend}`);

      const request = restclient({
        timeout: 1000,
      });

      return request
        .post(`${this.url}/internal/tracks`, {
          context: buildFlowStarterContext(),
          headers: { 'Content-Type': 'application/json' },
          data: { tracks: [trackToSend] },
        })
        .then((res) => {
          log.debug('postToMelidata done');
          return res;
        });
    }

    log.info('Not posting to melidata api. Enable it in configuration');
    return Promise.resolve();
  }
}

module.exports = MelidataRecordService;
