// frontend-page/src/constants.js
const PAGES_PATH = './app/nordic-pages';
const PAGE_COMPONENTS_PATH = '**/ui-components/**';
const PAGE_TESTS_DIRECTORY_PATH = '**/__test?(s)__/**';
const PAGE_POLYFILLS_FILE = 'polyfills.js';

const FILES_IGNORED_IN_DISCOVERY = [PAGE_COMPONENTS_PATH, PAGE_TESTS_DIRECTORY_PATH, PAGE_POLYFILLS_FILE];

const hooksFilesIdentifier = {
  VIEW_SPECIFIC: 'hooks.server',
  BASE: '_hooks.server',
};

const discoveryFileExtensions = {
  SCRIPT: ['tsx', 'ts', 'jsx', 'js'],
  STYLE: ['scss', 'css'],
};

const environments = {
  DEV: 'development',
  PROD: 'production',
  TEST: 'test',
};
const possibleDeviceTypes = ['desktop', 'mobile', 'tablet', 'webview'];
const discoveryDeviceTypes = [...possibleDeviceTypes, 'base'];
const possibleAdaptiveOptions = {
  both: 'both',
  scripts: 'scripts',
  styles: 'styles',
};

const warnings = {
  MELIDATA: 'You are trying to use melidata but you are not specifying a path. '
  + 'This is required for melidata to work. Melidata track will not be '
  + 'rendered to the page',
  ANALYTICS: 'You are trying to use analytics but you are not specifying a '
  + 'section. This is required for analytics to work. Analytics track will '
  + 'not be rendered to the page',
  ROUTE_DISCOVERY: 'The following paths could not be processed correctly:',
};

module.exports = {
  PAGES_PATH,
  PAGE_COMPONENTS_PATH,
  PAGE_POLYFILLS_FILE,
  PAGE_TESTS_DIRECTORY_PATH,
  FILES_IGNORED_IN_DISCOVERY,
  hooksFilesIdentifier,
  discoveryDeviceTypes,
  discoveryFileExtensions,
  environments,
  possibleDeviceTypes,
  possibleAdaptiveOptions,
  warnings,
};
