// nav-cart/src/widget/helpers/Layer.js
/* global document */
const positioner = require('./positioner');

class Layer {
  constructor(options) {
    this.options = options;

    this.el = this.constructEl();

    this.show();

    this.hide();
  }

  constructEl() {
    const el = document.createElement('div');
    el.className = 'ui-cart-feedback';
    // Set an initial state of the component
    el.innerHTML = this.options.template;

    // Set hidden as default. To set shown by default just execute .show()
    // method right after instantiation of the widget.
    el.hidden = true;
    el.setAttribute('aria-hidden', true);

    // Append at the end of the <body>
    document.body.appendChild(el);

    return el;
  }

  hide() {
    const that = this;
    setTimeout(() => {
      that.el.hidden = true;
    }, this.options.hideDelay);
  }

  /**
   * Shows the container.
   * @method
   */
  show() {
    this.el.hidden = false;
    this.el.setAttribute('aria-hidden', false);
    positioner(this.el, this.options.element);
  }
}

module.exports = Layer;
