// ml-web-navigation/src/navigation/header/components/user-menu/mobile/logged-in/welcome/index.js
const React = require('react');
const PropTypes = require('prop-types');
const { Image } = require('frontend-image');
const Loyalty = require('../loyalty');
const Avatar = require('../avatar');
const Melicoins = require('../melicoins');

const namespace = 'nav-header-menu-mobile';
const Welcome = (props) => {
  const {
    hasLoyalty,
    loyalty,
    avatar,
    title,
    firstName,
    nickname,
    email,
    mercadoPago,
  } = props;
  const className = hasLoyalty ? `${namespace}-with-loyalty` : undefined;

  const hasLevel6Drawer = hasLoyalty
    && loyalty
    && loyalty.drawer
    && loyalty.drawer.image
    && loyalty.drawer.alt
    && loyalty.drawer.link;

  const styleWithLvl6Pill = {
    height: 'auto',
  };

  const stylesDiv = {
    width: '100%',
    display: 'flex',
    justifyContent: 'flex-start',
    alignContent: 'flex-start',
    marginLeft: '55px',
  };

  const stylesImg = {
    maxWidth: '100%',
    maxHeight: '46px',
  };
  const stylesBoxHeader = {
    minHeight: '64px',
  };
  return (
    <div
      id={`${namespace}-user-info`}
      className={className}
      style={styleWithLvl6Pill}
    >
      <div style={stylesBoxHeader}>
        <Avatar {...avatar} hasLoyalty={hasLoyalty} loyalty={loyalty} />
        <div id="nav-header-user-greetings">
          {title || firstName || nickname}
        </div>
        <Loyalty hasLoyalty={hasLoyalty} email={email} {...loyalty} />
        {hasLevel6Drawer && (
          <div className={className} style={stylesDiv}>
            <a href={loyalty.drawer.link}>
              <Image
                src={loyalty.drawer.image}
                style={stylesImg}
                alt={loyalty.drawer.alt}
              />
            </a>
          </div>
        )}
      </div>
      {mercadoPago && <Melicoins mercadoPago={mercadoPago} />}
    </div>
  );
};

Welcome.propTypes = {
  hasLoyalty: PropTypes.bool.isRequired,
  loyalty: PropTypes.object.isRequired,
  avatar: PropTypes.object.isRequired,
  title: PropTypes.string,
  firstName: PropTypes.string,
  email: PropTypes.string,
  nickname: PropTypes.string,
  mercadoPago: PropTypes.object,
};

Welcome.defaultProps = {
  title: null,
  firstName: null,
  nickname: '',
  email: null,
};

module.exports = Welcome;
