// frontend-device_detection/src/browser.js
/**
 * Module dependencies
 */
const uaParser = require('ua-parser-js');
const semver = require('semver');
const defineSupport = require('./support');
const defineOutdated = require('./outdated');

/**
 * Expose device
 * Uses headers, user agent and cookies to find support of specific features.
 * @see mercadolibre/frontend-rules/subdomains/frontend.conf
 * @returns
 * {
 *   support, // Object
 *   name, // String
 *   version, // String
 *   userAgent, // String
 * }
 */
module.exports = (headers, lowEnd, userAgentKey = 'user-agent') => {
  const agent = uaParser(headers[userAgentKey]);
  const parsedVersion = semver.coerce(agent.browser.version);
  agent.major = parsedVersion ? parsedVersion.major : 0;

  return {
    support: defineSupport(headers, agent),
    name: agent.browser.name?.toLowerCase(),
    version: agent.browser.version,
    major: agent.major,
    minor: parsedVersion ? parsedVersion.minor : 0,
    patch: parsedVersion ? parsedVersion.patch : 0,
    userAgent: agent.ua,
    device: agent.device,
    outdated: defineOutdated(headers, lowEnd),
  };
};
