// frontend-device_detection/src/regexp.js
/**
 * Exports regular expressions to determine device types
 */
exports.mobile = /.*(blackberry|android|iphone|ipod|symb|googlebot-mobile|Playstation.Vita|kindle|symbian|windows.phone|BB10|Mobile.*Firefox|Nokia[0-9][0-9]|Nokia[A-Z][0-9]|LG-[A-Z][0-9]|SAMSUNG-GT|SPICE|Opera.Mini).*/i;

exports.tablet = /.*(ipad|Silk.1.0.22.79|android(?!(.*mobile|.*kindle|.*A2107A|.*HP.7|.*opera.mini|.*HP.Slate.7|.*B1-710|.*GT-P3100|.*GT-P1000|.*IdeaTabA1000|.*gt-P3110|.*gt-P3113|.*NX008HD8G|.*ME172V|.*ME173X|.*AT300|.*MediaPad.7.Lite|.*B1-A71|.*A200|.*A100|.*SM-T210R|.*SM-T211|.*SM-T217S|.*SGPT12|.*SM-T310|.*K00U.Build))(?!(.*MMS-V))|xoom|sch-i800|playbook|tablet|gt-P5100|gt-P7500|gt-P7510|kindle.fire|.*KFOT.Build|.*KFTT.Build|.*KFJWI.Build|.*KFJWA.Build).*/i;

exports.lowEnd = /.*(SPICE|BB10|SAMSUNG-GT-C3322|SAMSUNG-GT-E2200|SAMSUNG-GT-S5222|SAMSUNG-GT-S3570|S3570XILH1|blackberry|weblight|symbian|opera.mini|windows.phone|MSIE\s(5|6|7|8|9|10)|AppleWebKit\/(4|5(1|2|3[0-6]))|Firefox\/(([1-9]|1[0-9]|2[0-3])\.)|Nokia[0-9][0-9]|Nokia[A-Z][0-9]|LG-C[0-9]|LG-T[0-9]).*/i;

exports.pciComliant = /(MSIE\s(5|6|7|8|9|10))|(IEMobile\/[1,7-9]0?\.)|(Chrome\/([1-9]|[1][0-9]|2[01])\.)|(Firefox\/([1-9]|[1][0-9]|2[0-3])\.)|(OS\s[34]_)|(Version\/[1-6]\.\d+.\d+\sSafari)|(Version\/[1-6]\.\d+\sMobile\/.+Safari)|(Android\s[1-4].\d+(.\d+)?.+Version)|(Opera\sMini\/[1-4]\.)|(Opera\s1?[0-9]\.)|(Opera\/.+Version\/([1-9]\.|10\.|11\.|12\.(0|1[1-5])))/;

exports.nativeApp = /^([a-zA-Z]+)-(iOS|Android|AndroidNoPS|SmartPOS)?(?:\/|%2F)((?:\d+\.){2}\d+)/;

exports.huaweiNoGPS = /(TAS|LIO|ELS|JNY|ANA|CDY|ART|MED|AQM|CND|DVC|Kobe2|MRX|KOB2|JAD|ABR|BAL|JLN|NAM|JLN|RTE|TET|MXW|LRA|OXF|EBG|BMH|VNE|SPL|DRA|FRL|PPA|NEN|WKG)/;

/**
 * Devices with no complete SameSite support
 *
 * IE: *
 * IEMobile: *
 * Edge: <= 17
 * Firefox: <= 59
 * Chrome: <= 50
 * Safari: <= 13 && macOS < 10.15
 * iOS Safari: <= 12
 * Android Browser: *
 * Opera: <= 38
 * Opera Mini: *
 * Samsung Internet: <= 4
 * UC Browser: *
 */
exports.sameSiteSupport = /(MSIE)|(IEMobile)|(Edge?\/([0-9]|(1[0-7]))\.)|(Firefox\/([1-9]|[1-5][0-9])\.)|(Chrome\/([1-9]|([1-5][0-9])|6[0-6])\.)|(OS\sX\s10_([0-9]|(1[0-4]))_.*Version\/\d+(\.\d+)+\sSafari)|(Version\/([1-9]|(1[0-2]))(\.\d+)+\sMobile\/.+Safari)|(OS\s1([0-1]|2_[0-4]).*\slike\sMac\sOS\sX.+Mobile)|(Android\s[1-4](.\d+)+.+Version)|(Dalvik\/1\.\d+.+Android)|(Opera[\s/+]1?[0-9]\.)|(OPR\/(([1-2][0-9])|(3[0-8])))|(SamsungBrowser\/[1-4]\.)|(UC[\s+]?Browser\/?\d+)/i;

/**
 * Browsers without support
 * About 18 months ago
 *
 * IE: ALL VERSIONS ARE OUTDATED
 * IEMobile: ALL VERSIONS ARE OUTDATED
 * Edge: <= Engine 16 - Release 41 https://en.wikipedia.org/wiki/Microsoft_Edge
 * Firefox: <= 62 https://www.mozilla.org/en-US/firefox/releases/
 * Chrome: <= 66 https://en.wikipedia.org/wiki/Google_Chrome_version_history
 * Safari: <= 10 https://en.wikipedia.org/wiki/Safari_version_history
 * iOS Safari: <= 10
 * Android Browser: *
 * Opera: *
 * Opera Mini: *
 * Samsung Internet: *
 * UC Browser: *
 */
exports.outdatedSupport = /(MSIE)|(Trident)|(IEMobile)|(WPDesktop)|(Edge?\/([0-9]|(1[0-6]))\.)|(Firefox\/([0-9]|[1-5][0-9]|6[0-2])\.)|(Chrome\/([0-9]|([1-5][0-9])|6[0-6])\.)|(Version\/([1-9]|10)\.\d+(\.\d+)?.Safari)|(Version\/([1-9]|(10))(\.\d+).+Mobile\/.+Safari)|(CriOS\/([0-9]|([1-5][0-9])|6[0-6])\..+Mobile\/.+Safari)/i;
