// frontend-device_detection/src/support.js
/**
 * Module dependencies
 */
const checkWebP = require('./webp');
const checkPrefetch = require('./prefetch');
const checkPreconnect = require('./preconnect');
const checkPushNotifications = require('./push-notifications');
const checkServiceWorkers = require('./service-workers');
const checkBrotli = require('./brotli');
const checkSameSite = require('./same-site');

/**
 * Expose the browser support capabilities
 * Uses headers, user agent and cookies to detect the support of specific features.
 * @see mercadolibre/frontend-rules/subdomains/frontend.conf
 * @returns
 * {
 *   webp,     // Boolean
 *   prefetch, // Boolean
 *   pushNotifications, // Boolean
 *   serviceWorkers, // Boolean
 *   preconnect, // Boolean
 * }
 */
module.exports = (headers, agent) => ({
  webp: checkWebP(headers.accept, agent),
  prefetch: checkPrefetch(agent),
  preconnect: checkPreconnect(agent),
  pushNotifications: checkPushNotifications(agent),
  serviceWorkers: checkServiceWorkers(agent),
  brotli: checkBrotli(headers['accept-encoding']),
  sameSite: checkSameSite(headers),
});
