// frontend-melidata/config/melidata.js
const PRODUCTION_ENV = process.env.NODE_ENV === 'production';
const MELIDATA_SDK_PRODUCTION = process.env.MELIDATA_SDK_PRODUCTION === 'true';
const MELIDATA_ON_PRODUCTION = (PRODUCTION_ENV || MELIDATA_SDK_PRODUCTION);
const MELIDATA_USE_AUDIENCES_INTEGRATION = process.env.MELIDATA_USE_AUDIENCES_INTEGRATION === 'true';

module.exports = {
  MELIDATA_SDK_PRODUCTION,
  MELIDATA_ON_PRODUCTION,
  MELIDATA_USE_AUDIENCES_INTEGRATION,
  MAJOR_JS_VERSION: '3',
  DO_POST: MELIDATA_ON_PRODUCTION,
  JS_SCRIPT_NAME: MELIDATA_ON_PRODUCTION ? 'melidata.min.js' : 'melidata_test.js',
  DEFAULT_SCRIPT_URL: 'http://http2.mlstatic.com/storage/melidata-js-sdk/js/#mayorJSVersion#/last/#jsScriptName',
  DEFAULT_SCRIPT_URL_SECURE: 'https://http2.mlstatic.com/storage/melidata-js-sdk/js/#mayorJSVersion#/last/#jsScriptName',
  DEFAULT_PIXEL_URL: '//api.mercadolibre.com/pixel.gif',
  DEFAULT_RELOAD_INTERVAL: 60000,
  CONFIG_PROPERTIES_URL: 'http://internal.mercadolibre.com/melidata-js-sdk/static/js/#mayorJSVersion#/last/config.properties',
  DEFAULT_MELIDATA_ML_ENDPOINT: PRODUCTION_ENV ? 'http://internal.mercadolibre.com' : 'https://internal-api.mercadolibre.com',
  DEFAULT_MELIDATA_MP_ENDPOINT: PRODUCTION_ENV ? 'http://internal.mercadopago.com' : 'https://internal-api.mercadopago.com',
  DEFAULT_DATADOG_TRACKING: false,
};
