// mp-web-navigation/src/navigation/header/components/user-menu/navigation/landing-navigation/mobile/index.js
const React = require('react');
const PropTypes = require('prop-types');
const getDataBySite = require('../../../../../../utils/data-management');
const ListContent = require('../list-content');

const LandingNavigationMobile = (props) => {
  const { enabled, structure, translations, urls, collapsibleItemLabel, queryParams } = props;
  const data = getDataBySite(enabled, structure, translations, urls, queryParams);

  return (
    <div className="ln-list-container">
      {
        Object.keys(data).map((key) => (
          <React.Fragment key={`listcontent-key-${key}`}>
            <ListContent
              id={key}
              header={data[key].header}
              listItems={data[key].items}
              texts={collapsibleItemLabel}
            />
            <hr className="ln-list-container__separator" />
          </React.Fragment>
        ))
      }
    </div>
  );
};

LandingNavigationMobile.defaultProps = {
  queryParams: {},
};

LandingNavigationMobile.propTypes = {
  siteId: PropTypes.string.isRequired,
  structure: PropTypes.shape({}).isRequired,
  translations: PropTypes.shape({}).isRequired,
  urls: PropTypes.shape({}).isRequired,
  collapsibleItemLabel: PropTypes.shape({}).isRequired,
  queryParams: PropTypes.shape({}),
};

module.exports = LandingNavigationMobile;
