// mp-web-navigation/src/navigation/footer/plus/index.js
const React = require('react');
const PropTypes = require('prop-types');
const Logo = require('../components/logo');
const NavSection = require('../components/navigation');
const Social = require('../components/social');
const Copyright = require('../components/copyright');

const FooterPlus = ({ logo, navSections, reclamationBook, social, copyright, siteId }) => (
  <footer role="contentinfo" className="nav-footer-mp">
    <div className="nav-footer-mp__navs">
      <Logo {...logo} />
      {navSections.navHeaderSections.map((navSection) => (
        <NavSection key={`navsection-${navSection.id}`} {...navSection} />
      ))}
      <NavSection {...navSections.navAboutSection} siteId={siteId} reclamationBook={reclamationBook} />
    </div>
    <div className="nav-footer-mp__last">
      <NavSection {...navSections.navPerformanceSection} className="performance" />
      <section className="copyright-and-social">
        <Copyright {...copyright} />
        <Social {...social} />
      </section>
    </div>
  </footer>
);

FooterPlus.propTypes = {
  logo: PropTypes.shape({}).isRequired,
  navSections: PropTypes.shape({
    navHeaderSections: PropTypes.arrayOf(
      PropTypes.shape({}),
    ),
    navAboutSection: PropTypes.shape({}),
    navPerformanceSection: PropTypes.shape({}),
  }).isRequired,
  reclamationBook: PropTypes.shape({}),
  social: PropTypes.shape({}).isRequired,
  copyright: PropTypes.shape({}).isRequired,
  siteId: PropTypes.string.isRequired,
};

FooterPlus.defaultProps = {
  reclamationBook: null,
};

module.exports = FooterPlus;
