// ml-web-navigation/src/navigation/header/components/user-menu/desktop/logged-in/notifications-center/enhancement.js
/* eslint-disable no-script-url */
/* eslint-disable security/detect-non-literal-fs-filename */
const React = require('react');
const classNames = require('classnames');
const Icon = require('./icon');

const id = 'notificationsCenter';
const notificationsCenterWidget = 'notifications-widget';

const NotificationsCenterEnhancement = (url, props) => {
  const { domain } = props.platform;
  const notificationsUrl = `'https://www.${domain}/gz/notifications/:offset/:limit'`;
  const deleteUrl = `'https://www.${domain}/gz/notifications/:newsgroupId'`;
  const countUrl = `'https://www.${domain}/gz/notifications/badge'`;
  const settingsUrl = `'https://www.${domain}/gz/notifications/settings'`;

  const newUrl = { ...url,
    name: (<Icon text={url.name} />),
    link: { ...url.link,
      href: url.link.href,
      className: classNames('option-notifications', notificationsCenterWidget) },
    modelessBoxCallback: `
    window.layoutNotifications = new Notifications({
      trigger: document.querySelector('.${notificationsCenterWidget}'),
      hostname: mainDomain,
      fixed: ${props.fixed},
      bus: window.freya,
      siteId: '${props.platform.siteId}',
      url: ${notificationsUrl},
      deleteUrl: ${deleteUrl},
      countUrl: ${countUrl},
      credentials: 'true',
      isHover: true,
      settingsUrl: ${settingsUrl}
    });
    ` };

  return newUrl;
};

module.exports = {
  id,
  enhance: NotificationsCenterEnhancement,
};
