// ml-web-navigation/src/navigation/footer/components/sites/sic/index.js
const { getValue } = require('ml-navigation-config');
const { isExampleApp } = require('@frontend-navigations/commons/services/example-app');
const { TEST } = require('frontend-env');
const isMobile = require('../../../../utils/is-mobile');
const Tertiary = require('./tertiary');
const { name } = require('../../../../../../package.json');

const getUrlBase = (props) => {
  const uiNavigationVersion = getValue({
    key: 'uiNavigationVersion',
    req: props.req,
    platform: props.platform,
  });

  const baseUrl = isExampleApp(name) && !TEST
    ? 'http://localhost:8081'
    : 'https://http2.mlstatic.com';

  return `${baseUrl}/frontend-assets/ml-web-navigation/ui-navigation/${uiNavigationVersion}/mercadolibre`;
};

const doesApply = (props) => {
  const { platform, req } = props;

  const sicConfig = getValue({
    key: 'sic',
    req,
    platform,
  });

  return !!sicConfig;
};

const sicProps = (props) => {
  const { platform, req } = props;
  const sicConfig = getValue({
    key: 'sic',
    req,
    platform,
  });

  const urlBase = getUrlBase(props);

  return {
    link: {
      href: sicConfig.url,
      name: sicConfig.name,
      id: 'sic',
    },
    sic: {
      url: sicConfig.url,
      image: `${urlBase}/${sicConfig.sicImageName}`,
      bigImage: `${urlBase}/${sicConfig.sicImage2xName} 2x`,
      alt: sicConfig.altSic, // 'SIC - Industria y comercio',
    },
    pum: {
      url: sicConfig.urlPum,
      image: `${urlBase}/${sicConfig.pumImageName}`,
      bigImage: `${urlBase}/${sicConfig.pumImage2xName} 2x`,
      alt: sicConfig.altPum, // 'Precio por unidad de medida',
      width: '200',
      height: '34',
    },
    isMobile: isMobile(props),
  };
};

module.exports = {
  doesApply,
  Tertiary,
  props: sicProps,
};
