// cbt-web-navigation/src/navigation/header/components/user-menu/logged-in/index.js
const React = require('react');
const PropTypes = require('prop-types');
const UserMenuWidget = require('../../../../widgets/user-menu');
const LoggedInCurrentSite = require('./current-site/index');
const loggedInCurrentSiteProps = require('./current-site/props');
const LoggedInOptions = require('./options');
const loggedInOptionsProps = require('./options/props');
const types = require('../../../../types');

const nameSpace = 'nav-header-cbt';

const LoggedInMenu = (props) => {
  const { userName, currentSiteId } = props;

  return (
    <div id={`${nameSpace}__logged`} className={`${nameSpace}-body`}>
      <div className={`${nameSpace}__trigger`}>
        <div className={`${nameSpace}__trigger-element`}>{userName}</div>
        {currentSiteId && <LoggedInCurrentSite {...loggedInCurrentSiteProps(props)} />}
        <i className={`${nameSpace}__trigger-element ${nameSpace}__chevron`} />
      </div>
      <LoggedInOptions {...loggedInOptionsProps(props)} />
      <UserMenuWidget />
    </div>
  );
};

LoggedInMenu.defaultProps = {
  currentLogisticType: '',
  cbtInfo: null,
  currentSiteId: '',
  sites: null,
  logisticTypes: null,
};

LoggedInMenu.propTypes = {
  cbtInfo: PropTypes.shape(types.cbtInfoType),
  userName: PropTypes.string.isRequired,
  currentLogisticType: PropTypes.string,
  currentSiteId: PropTypes.string,
  sites: PropTypes.objectOf(PropTypes.shape(types.siteType)),
  logisticTypes: PropTypes.shape(types.logisticType),
  links: PropTypes.objectOf(PropTypes.shape(types.linkType)).isRequired,
};

module.exports = LoggedInMenu;
