// ml-web-navigation/src/navigation/header/components/nav-menu/index.js
const React = require('react');
const PropTypes = require('prop-types');
const Link = require('../nav-link');

const namespace = 'nav-menu';

const NavMenu = ({ links, show, linkTags }) => show && (
  <div className={namespace}>
    <ul className={`${namespace}-list`}>
      {
        links.map(link => (
          link && (
            <li
              key={link.id}
              className={`${namespace}-item`}
            >
              <Link
                {...link}
                type="nav-link"
                linkTag={linkTags[link.id]}
              />
            </li>
          )
        ))
      }
    </ul>
  </div>
);


NavMenu.propTypes = {
  links: PropTypes.arrayOf(PropTypes.shape({
    link: PropTypes.shape({
      href: PropTypes.string,
      className: PropTypes.string,

    }).isRequired,
    name: PropTypes.string.isRequired,
  })).isRequired,
  show: PropTypes.bool.isRequired,
  linkTags: PropTypes.object, // eslint-disable-line react/forbid-prop-types
};

NavMenu.defaultProps = {
  linkTags: {},
};

module.exports = NavMenu;
