// @meli/mobile-webkit/src/events/event-emitter.ts
// xq generar a traves de un doc
// el evento se llamaria por cada metodo de un topic?

export class EventEmitter implements EventTarget {

    addEventListener(type: string, callback: EventListenerOrEventListenerObject, options?: boolean | AddEventListenerOptions): void {
        window.addEventListener(type, callback, options);
    }

    removeEventListener(type: string, callback: EventListenerOrEventListenerObject, options?: boolean | EventListenerOptions): void {
        window.removeEventListener(type, callback, options);
    }

    dispatchEvent(event: Event): boolean {
        return window.dispatchEvent(event);
    }

    
    on(type: string, callback: EventListenerOrEventListenerObject, options?: boolean | AddEventListenerOptions): void {
        this.addEventListener(type, callback, options);
    }

    off(type: string, callback: EventListenerOrEventListenerObject, options?: boolean | EventListenerOptions): void {
        this.removeEventListener(type, callback, options);
    }

    emit(event: Event): boolean {
        return this.dispatchEvent(event);
    }

}
