// ml-web-navigation/src/navigation/css/index.js
const React = require('react');
const propTypes = require('prop-types');
const CommonsCss = require('@frontend-navigations/commons/lib/css-component');
const {
  DESKTOP,
  TABLET,
  MOBILE,
} = require('@frontend-navigations/commons/constants/device-types');
const {
  INTERNAL,
  FULL,
} = require('../../constants/navigation-types');
const {
  NAVIGATION_INTERNAL,
  NAVIGATION_MOBILE,
  NAVIGATION_TABLET,
  NAVIGATION_DESKTOP,
  NAVIGATION,
  NAVIGATION_LARGE,
} = require('../../constants/css-files');

const deviceTypes = {
  [MOBILE]: [{
    name: NAVIGATION_MOBILE,
  }],
  [TABLET]: [{
    name: NAVIGATION_TABLET,
  }],
  [DESKTOP]: [{
    name: NAVIGATION_DESKTOP,
    media: '(min-width: 1024px)',
  }, {
    name: NAVIGATION_MOBILE,
    media: '(max-width: 1023px)',
  }],
};

const Css = ({
  navigationType,
  deviceType,
  browser,
  firstRequestInlineStyles,
  platformId,
  siteId,
  embedCss,
  feConfigEnvironment,
  cdn,
  configGetter,
  configEnvironment,
  addConfigUpdateHandler,
}) => {
  // navigation-internal__small.css
  const baseFiles = (navigationType === INTERNAL && [{ name: NAVIGATION_INTERNAL }])
    // navigation-mobile.css navigation-desktop.css navigation-tablet.css
    || deviceTypes[deviceType] // eslint-disable-line security/detect-object-injection
    // navigation.css
    || [{ name: NAVIGATION }];

  const loadLargeCss = browser && browser.name === 'ie' && parseInt(browser.major, 10) < 9;
  const files = baseFiles
    .concat(loadLargeCss ? { name: NAVIGATION_LARGE } : [])
    .map(style => ({
      ...style,
      firstRequestInlineStyles,
    }));

  return (
    <CommonsCss
      files={files}
      platformId={platformId}
      siteId={siteId}
      embedCss={embedCss}
      feConfigEnvironment={feConfigEnvironment}
      navigationType={navigationType}
      cdn={cdn}
      configGetter={configGetter}
      configEnvironment={configEnvironment}
      addConfigUpdateHandler={addConfigUpdateHandler}
      appName="ml-web-navigation"
    />
  );
};

Css.propTypes = {
  firstRequestInlineStyles: propTypes.bool,
  navigationType: propTypes.string,
  deviceType: propTypes.string,
  browser: propTypes.shape({
    name: propTypes.string,
    major: propTypes.number,
  }),
  platformId: propTypes.string,
  siteId: propTypes.string,
  embedCss: propTypes.bool,
  feConfigEnvironment: propTypes.string,
  cdn: propTypes.shape({
    link: propTypes.string,
  }),
  configGetter: propTypes.func,
  configEnvironment: propTypes.string,
  addConfigUpdateHandler: propTypes.func,
};

Css.defaultProps = {
  firstRequestInlineStyles: false,
  navigationType: FULL,
  browser: null,
  deviceType: null,
  platformId: 'ML',
  siteId: 'MLA',
  embedCss: false,
  feConfigEnvironment: null,
  cdn: null,
  configGetter: null,
  configEnvironment: null,
  addConfigUpdateHandler: null,
};

module.exports = Css;
