// mp-web-navigation/src/navigation/header/components/user-menu/navigation/landing-navigation/desktop/index.js
const React = require('react');
const PropTypes = require('prop-types');
const getDataBySite = require('../../../../../../utils/data-management');
const ContainerEffects = require('../../../../../../widgets/container-effects');
const ListContent = require('../list-content');
const NavContainer = require('../nav-container');

const LandingNavigationDesktop = (props) => {
  const { enabled, structure, translations, urls, collapsibleItemLabel, queryParams } = props;
  const data = getDataBySite(enabled, structure, translations, urls, queryParams);

  const navLinks = Object.keys(data).map((navLink) => ({
    id: navLink,
    text: data[`${navLink}`].label,
  }));

  /* eslint-disable react/no-array-index-key */
  return (
    <div className="landing-list" id="landing-navigation-container--desktop">
      <NavContainer items={navLinks} />
      {
        navLinks.map((item, index) => (
          <div
            key={`navlink-key${index}`}
            id={item.id}
            data-id={item.id}
            className="ln-list-container ln-list-container--close"
          >
            <div className="ln-list-container__triangle" />
            <ListContent
              id={item.id}
              header={data[`${item.id}`].header}
              listItems={data[`${item.id}`].items}
              texts={collapsibleItemLabel}
            />
          </div>
        ))
      }
      <ContainerEffects />
    </div>
  );
};

LandingNavigationDesktop.defaultProps = {
  namespace: '',
  queryParams: {},
};

LandingNavigationDesktop.propTypes = {
  siteId: PropTypes.string.isRequired,
  structure: PropTypes.shape({}).isRequired,
  translations: PropTypes.shape({}).isRequired,
  urls: PropTypes.shape({}).isRequired,
  collapsibleItemLabel: PropTypes.shape({}).isRequired,
  namespace: PropTypes.string,
  queryParams: PropTypes.shape({}),
};

module.exports = LandingNavigationDesktop;
