// vis-web-navigation/src/navigation/widgets/searchbox/index.js
const React = require('react');
const PropTypes = require('prop-types');
const Script = require('frontend-script');

const SearchboxWidget = ({
  platformId,
  siteId,
  baseUrl,
  version,
  searchEndpoint,
  categoryEndpoint,
  autosuggestDomain,
  isUserLoggedIn,
  siteDomain,
  deviceType,
}) => {
  const url = `${baseUrl}/ui/searchbox/${version}/index.js`;
  return (
    <Script src={url}>{
      `
      if (window.Searchbox) {
        new window.Searchbox('.nav-search-input', {
          platformId: '${platformId}',
          siteId: '${siteId}',
          siteDomain: '${siteDomain}',
          limit: 6,
          searchEndpoint: '${searchEndpoint}',
          categoryEndpoint: '${categoryEndpoint}',
          categoryCheckbox: document.getElementById('categorySearch'),
          loggedIn: ${isUserLoggedIn},
          ${autosuggestDomain ? `autosuggestDomain: '${autosuggestDomain}',` : ''}
          bus: window.freya,
          deviceType: '${deviceType}'
        });
      }
    `
}
    </Script>
  );
};

SearchboxWidget.propTypes = {
  baseUrl: PropTypes.string.isRequired,
  platformId: PropTypes.string.isRequired,
  siteId: PropTypes.string.isRequired,
  searchEndpoint: PropTypes.string.isRequired,
  categoryEndpoint: PropTypes.string,
  isUserLoggedIn: PropTypes.bool.isRequired,
  siteDomain: PropTypes.string.isRequired,
  deviceType: PropTypes.string.isRequired,
  autosuggestDomain: PropTypes.string,
  version: PropTypes.string,
};

SearchboxWidget.defaultProps = {
  version: '2.2.3',
  categoryEndpoint: null,
  autosuggestDomain: null,
};

module.exports = SearchboxWidget;
