// frontend-melidata/src/experiments/entry.js
class Entry {
  constructor(fetchDate, expiresIn, stalesIn, data) {
    this._fetchDate = fetchDate;
    this._expiresIn = expiresIn;
    this._stalesIn = stalesIn;
    this._data = data;
  }

  getFetchDate() { return this._fetchDate; }

  getExpiresIn() { return this._expiresIn; }

  getStalesIn() { return this._stalesIn; }

  getData() { return this._data; }

  refresh(e) {
    this._fetchDate = e._fetchDate;
    this._expiresIn = e._expiresIn;
    this._stalesIn = e._stalesIn;
    this._data = e._data;
  }

  addItem(item) {
    if (item && item.name) {
      this._data[item.name] = item;
    }
  }

  clear() {
    this._fetchDate = null;
    this._expiresIn = null;
    this._stalesIn = null;
    this._data = null;
  }

  toString() {
    const {
      _fetchDate, _expiresIn, _stalesIn, _data,
    } = this;

    return JSON.stringify({
      fetchDate: _fetchDate,
      expiresIn: _expiresIn,
      stalesIn: _stalesIn,
      data: _data,
    }, null, 2);
  }
}

module.exports = Entry;
