// pi-web-navigation/src/navigation/header/components/user-menu/desktop/guest/props.js
const { mapLinksToUrls } = require('../../../../../utils/links');
const { isLite } = require('../../../../../../layout-type-detection');

// TODO: move to conf
const login = 'login';
const sell = 'sell';
const help = 'helpMain';
const registration = 'registration';

const guestUserMenuProps = (props) => {
  const { navigationUrls: urls, actualLocation, type } = props;

  const userMenuLinks = [];

  if (isLite(type)) {
    userMenuLinks.push(help);
  } else {
    userMenuLinks.push(registration);
    userMenuLinks.push(login);
    userMenuLinks.push(help);
    userMenuLinks.push(sell);
  }

  const links = mapLinksToUrls({
    links: userMenuLinks,
    urls,
    currentUrl: actualLocation,
    originalProps: props,
    hash: 'nav-header',
    className: (link) => {
      switch (link) {
        case registration:
          return 'option-register';
        case help:
          return 'option-help';
        case sell:
          return 'option-publish';
        default:
          return `option-${link}`;
      }
    },
  });

  return {
    userLinks: {
      links,
    },
  };
};

module.exports = guestUserMenuProps;
