// pi-web-navigation/src/navigation/footer/props.js
/* eslint-disable max-len */
const copyrightProps = require('./components/copyright/props');
const baseFooterProps = require('./components/base-footer/props');
const footerAccessProps = require('./components/footer-access/props');
const footerLinksProps = require('./components/footer-links/props');
const userMenuProps = require('./components/user-menu/props');
const footerNavigationProps = require('./components/footer-navigation/props');
const honeypathTrap = require('./components/honeypath-trap/props');

const footerProps = (props) => {
  const baseCopyrightProps = copyrightProps(props);
  const basefooterProps = baseFooterProps(props);

  /**
   * PI's footer needs the copyright and the extra company name to be in the same line.
   * In order to reuse the component, we change the `companyName` to add that `extraCompanyName` and remove
   * the `extraCompanyName` from the baseFooterProps.
   */
  baseCopyrightProps.companyName = `${baseCopyrightProps.companyName} ${basefooterProps.extraCompanyName || ''}`;
  basefooterProps.extraCompanyName = null;

  return {
    type: props.type,
    copyright: baseCopyrightProps,
    baseFooter: basefooterProps,
    userMenu: userMenuProps(props),
    footerAccess: footerAccessProps(props),
    footerNavigation: footerNavigationProps(props),
    isDesktop: props.device.desktop || false,
    footerLinks: footerLinksProps(props),
    honeypathTrap: honeypathTrap(props),
  };
};

module.exports = footerProps;
