// frontend-layout/config/head-config.js
const CDNS = {
  default: {
    name: 'default',
    link: 'https://http2.mlstatic.com',
    cssFileSuffix: '',
  },
  akamaiQuic: {
    name: 'akamaiQuic',
    link: 'https://mlstaticquic-a.akamaihd.net',
    cssFileSuffix: '-akamai-quic',
  },
  assets: {
    name: 'assets',
    link: 'https://assets.mlstatic.com',
    cssFileSuffix: '-assets',
  },

  /*
  'akamai': { name: 'akamai', link: 'https://akamai.mlstatic.com', cssFileSuffix: '' },
  'century': { name: 'century', link: 'https://century.mlstatic.com', cssFileSuffix: '' },
  'cloudfront': { name: 'cloudfront', link: 'https://cloudfront.mlstatic.com', cssFileSuffix: '' },
  'http2': { name: 'http2', link: 'https://http2.mlstatic.com', cssFileSuffix: '-http2' },
  'verizon': { name: 'verizon', link: 'https://verizon.mlstatic.com', cssFileSuffix: '' },
  */
};

function getCDNProvider(cdnName) {
  return cdnName && CDNS[cdnName] ? CDNS[cdnName] : CDNS.default;
}

function cdnReplace(url, cdnName) {
  return url && url.replace && cdnName && CDNS[cdnName] ? url.replace(CDNS.default.link, CDNS[cdnName].link) : url;
}

module.exports = { getCDNProvider, CDNS, cdnReplace };
