// frontend-i18n/src/I18n.js
const i18nJsxCreator = require('./i18n-jsx-creator');
const I18nBase = require('./I18nBase');

/**
 * I18n
 */
class I18n extends I18nBase {
  constructor(options = {}) {
    super(options);
    this.jsx = {
      gettext: (text, settings) => {
        if (!settings) return null;

        const translation = this.getTranslation(text);
        return i18nJsxCreator.create(translation, settings);
      },
      ngettext: (text, nText, settings) => {
        if (!settings) return null;

        const translation = this.getPluralTranslation(text, nText, settings.quantity);
        return i18nJsxCreator.create(translation, settings);
      },
      pgettext: (msgctxt, text, settings) => {
        if (!settings) return null;

        const translation = this.getTranslation(text, msgctxt);
        return i18nJsxCreator.create(translation, settings);
      },
      npgettext: (msgctxt, text, nText, settings) => {
        if (!settings) return null;

        const translation = this.getPluralTranslation(text, nText, settings.quantity, msgctxt);
        return i18nJsxCreator.create(translation, settings);
      },
    };
  }
}

/**
 * Expose I18n
 */
module.exports = I18n;
