// @meli/mobile-webkit/src/commands/actions/actions_commands_impl.ts
import { ActionCommands } from "..";
import { MobileWebkit } from "../../mobile-webkit";
import { PickImageOptions } from "./pick_image_options";
import { PickImageResult } from "./pick_image_result";
import { ShareImageOptions } from "./share_image_options";

export class ActionsCommandsImpl implements ActionCommands {

    constructor(private mobileWebkit: MobileWebkit) { }

    shareImage(args: ShareImageOptions): Promise<string> {
        return this.mobileWebkit.executeNative({
            method: "share_image",
            args
        });
    }

    downloadImage(args: ShareImageOptions): Promise<string> {
        return this.mobileWebkit.executeNative({
            method: "download_image",
            args
        });
    }

    pickImage(args: PickImageOptions): Promise<PickImageResult> {
        return this.mobileWebkit.executeNative({
            method: "pick_image",
            args
        });
    }

    share(title: string, data: string): Promise<string> {
        return this.mobileWebkit.executeNative({
            method: "share",
            args: {
                title,
                data
            }
        });
    }
}
