// ml-web-navigation/src/navigation/header/components/user-menu/mobile/logged-in/avatar/index.js
const React = require('react');
const PropTypes = require('prop-types');
const Image = require('frontend-image');
const Script = require('frontend-script');
const Icon = require('./icon');
const readClientFile = require('../../../../../../utils/read-client-file');

const scriptContent = readClientFile('plus-mobile-upload-avatar.js');

const namespace = 'nav-header-mobile-avatar';
const Avatar = (props) => {
  const {
    hasPicture,
    hasLoyalty,
    lowDensityPictureUrl,
    highDensityPictureUrl,
    lazyLoad,
    loyalty,
    action,
    actionLabel,
  } = props;

  return (
    <form id={`${namespace}-form`} method="POST" action={action} encType="multipart/form-data">
      <label htmlFor={`${namespace}-upload`}>
        { hasLoyalty && <Icon {...loyalty} />}
        { hasPicture
          ? (<Image
              id="nav-header-mobile-user-avatar"
              src={lowDensityPictureUrl}
              height="58"
              width="58"
              lazyload={lazyLoad}
              srcSet={`${highDensityPictureUrl} 2x`}
              onload="off"
              alt={actionLabel}
          />)
          : <i id="nav-header-mobile-user-avatar" className="nav-icon-create-account-mobile" /> }
        <input
          id={`${namespace}-upload`}
          name="file"
          type="file"
          accept="image/jpeg,.jpg,image/gif,.gif,image/png,.png,.jpeg,image/webp,.webp"
        />
      </label>
      <Script>
        {scriptContent}
      </Script>
    </form>
  );
};

Avatar.propTypes = {
  hasPicture: PropTypes.bool,
  loyalty: PropTypes.object.isRequired,
  hasLoyalty: PropTypes.bool.isRequired,
  lowDensityPictureUrl: PropTypes.string,
  action: PropTypes.string.isRequired,
  highDensityPictureUrl: PropTypes.string,
  lazyLoad: PropTypes.string,
  actionLabel: PropTypes.string,
};

Avatar.defaultProps = {
  lowDensityPictureUrl: null,
  highDensityPictureUrl: null,
  hasPicture: false,
  lazyLoad: 'off',
  actionLabel: '',
};

module.exports = Avatar;
