// ml-web-navigation/src/navigation/header/components/searchbox/props.js
const widgetProps = require('../../../widgets/searchbox/props');
const isMobile = require('../../../utils/is-mobile');
const isDesktop = require('../../../utils/is-desktop');

const searchboxProps = (props) => {
  const {
    searchFocus,
    searchValue,
    configSchema,
    device,
    category,
    platform,
  } = props;

  const {
    searchInputLabel,
    searchPlaceholder,
    searchIconLabel,
    searchPlaceholderMobile,
    closeTitle,
    clearTitle,
  } = configSchema.i18n;

  return {
    defaultValue: searchValue,
    isFocus: searchFocus,
    inputLabel: searchInputLabel,
    placeholder: searchPlaceholder,
    placeholderMobile: searchPlaceholderMobile,
    iconLabel: searchIconLabel,
    isLowEnd: device.lowEnd,
    action: `https://www.${platform.domain}/jm/search`,
    category,
    isMobile: isMobile(props),
    isDesktop: isDesktop(props),
    widget: widgetProps(props),
    closeTitle: closeTitle || 'Cerrar', // TODO: Remove condition when 'Cerrar' comes from i18n
    clearTitle: clearTitle || 'Limpiar', // TODO: Remove condition when 'Limpiar' comes from i18n
  };
};

module.exports = searchboxProps;
