// mp-web-navigation/src/navigation/utils/link.js
const React = require('react');
const PropTypes = require('prop-types');
const tracking = require('../../constants/tracking');
const getIcon = require('../icons/getIcon');

const Link = ({ id, className, link, target, showIcon, icon, name, ...rest }) => {
  const Icon = icon && getIcon(icon);

  return (
    <a
      className={`${className ? `${className} ${className}--${id.toLowerCase()}` : ''}`}
      href={link}
      target={target}
      {...rest}
      data-track-action={tracking.ACTION}
      data-track-id={id.replace(/_/g, '-')}
    >
      {showIcon ? (
        <div className={`${className ? `${className}-container` : ''}`.trim()}>
          <i className={`${className ? `${className}-icon` : ''}`.trim()}>
            <Icon />
          </i>
          <p className={`${className ? `${className}-text` : ''}`.trim()}>{name}</p>
        </div>
      )
        : name}
    </a>
  );
};

Link.propTypes = {
  id: PropTypes.string,
  link: PropTypes.string.isRequired,
  name: PropTypes.string.isRequired,
  className: PropTypes.string,
  target: PropTypes.string,
  icon: PropTypes.string,
  showIcon: PropTypes.bool,
};

Link.defaultProps = {
  id: '',
  className: null,
  target: null,
  showIcon: false,
  icon: null,
};

module.exports = Link;
