// frontend-melidata/src/utils/track-utils.js
const dateFormat = require('dateformat');
const crypto = require('crypto');
const jws = require('jws');

const DATETIME_FORMAT = 'yyyy-mm-dd\'T\'HH:MM:ss.LZ';
const ENCRYPTION_ALGORITHM = 'aes-256-ctr';
const IV_LENGTH = 16;
const ENCRYPTION_PASSWORD = 'f9bf78b9a18ce6d46a0cd2b0b86df9da';

const SIGNATURE_ALGORITHM = 'HS256';
const SIGNATURE_PASSWORD = '?E(H+MbQeThWmZq4t6w9z$C&F)J@NcRf';

function formatDate(date) {
  return dateFormat(date, DATETIME_FORMAT);
}

function addClaim(claims, key, value) {
  if (value) {
    claims[key] = value;
  }
}

function encryptData(track) {
  const host = track.application.server_hostname;
  const app = track.application.app_name;
  const pool = track.application.server_poolname;

  const securedFields = {
    application: {
      server_hostname: host,
      server_poolname: pool,
      app_name: app,
    },
  };

  const text = JSON.stringify(securedFields);

  const iv = crypto.randomBytes(IV_LENGTH);
  const cipher = crypto.createCipheriv(ENCRYPTION_ALGORITHM, ENCRYPTION_PASSWORD, iv);
  let encrypted = cipher.update(text);

  encrypted = Buffer.concat([encrypted, cipher.final()]);

  return `${iv.toString('hex')}:${encrypted.toString('hex')}`;
}

function signTrack(track, signKey) {
  const claims = {};
  addClaim(claims, 'uid', track.user.uid);
  addClaim(claims, 'user_id', track.user.user_id);
  addClaim(claims, 'site_id', track.application.site_id);
  addClaim(claims, 'platform', track.device.platform);

  let signatureKey = SIGNATURE_PASSWORD;

  if (signKey) {
    signatureKey = signKey;
  }

  return jws.sign({
    header: { alg: SIGNATURE_ALGORITHM },
    payload: claims,
    secret: signatureKey,
  });
}

/**
 * Apply security configs to track
 * @param track
 * @return
 */
function secureTrack(track, signKey) {
  track.application.secured = {};

  track.application.secured.encrypted = encryptData(track);
  track.application.secured.signature = signTrack(track, signKey);

  track.application.server_hostname = '';
  track.application.server_poolname = '';
  track.application.app_name = '';
}

module.exports = {
  formatDate,
  secureTrack,
};
