// frontend-navigations/index.js
/* eslint-disable security/detect-object-injection */
const MercadoLibreNavigation = require('ml-web-navigation');
const MercadoPagoNavigation = require('mp-web-navigation');
const CBTNavigation = require('cbt-web-navigation');
const MercadoShopsNavigation = require('ms-web-navigation');
const PortalInmobiliarioNavigation = require('pi-web-navigation');
const ClassiNavigation = require('vis-web-navigation');
const FirstPartyNavigation = require('first-party-web-navigation');

const navigations = {
  ML: {
    sites: {
      CBT: CBTNavigation,
    },
    default: MercadoLibreNavigation,
  },
  MP: {
    sites: {
      CBT: CBTNavigation,
    },
    default: MercadoPagoNavigation,
  },
  MS: {
    sites: {},
    default: MercadoShopsNavigation,
  },
  PI: {
    sites: {},
    default: PortalInmobiliarioNavigation,
  },
  TI: {
    sites: {},
    default: ClassiNavigation,
  },
  MT: {
    sites: {},
    default: ClassiNavigation,
  },
  TC: {
    sites: {},
    default: ClassiNavigation,
  },
  TL: {
    sites: {},
    default: ClassiNavigation,
  },
  TM: {
    sites: {},
    default: ClassiNavigation,
  },
};

/**
 * Returns a Navigation for the provided context
 * @param {Object} req Express request
 */
const create = (req) => {
  if (req.firstParty && req.firstParty.isFirstPartyMarketplaceFlow) {
    return FirstPartyNavigation;
  }
  const platformId = req.platform && req.platform.id;
  if (!platformId) {
    throw new Error('Platform parameter is required');
  }

  const platformData = navigations[platformId];
  if (!platformData) {
    return null;
  }

  const siteId = req.platform && req.platform.siteId;

  return (siteId && platformData.sites[siteId]) || platformData.default;
};

const navigationsHash = [
  MercadoLibreNavigation,
  CBTNavigation,
  MercadoPagoNavigation,
  MercadoShopsNavigation,
  PortalInmobiliarioNavigation,
  ClassiNavigation,
  FirstPartyNavigation,
].reduce((hash, navigation) => ({
  ...hash,
  [navigation.id]: navigation,
}), {});

const get = navigationId => navigationsHash[navigationId] || null;

module.exports = {
  create,
  get,
};
