// ragnar/config/csp.js
const {
  APP,
  APPLICATION,
  DEPT,
  PLATFORM,
  VERSION,
} = require('frontend-env');

/**
 * Generates a flow name for reports using the app name and version
 */
function generateFlowName() {
  const version = VERSION || 'dev';
  if (PLATFORM === 'fury') {
    const name = (APPLICATION || 'appdev').toLowerCase();
    return `app-${name}-${version}`;
  }

  const department = (DEPT || 'deptdev').toLowerCase();
  const name = (APP || 'appdev').toLowerCase();
  return `app-${department}-${name}-${version}`;
}

/**
 * Expose CSP default directives
 */
module.exports = {
  directives: {
    'script-src': [
      function addNonceCSP(req, res) {
        if (res.locals && res.locals.nonce) {
          return `'nonce-${res.locals.nonce}'`;
        }
        return null;
      },
      function addStrictDynamic(req, res) {
        if (res.locals && res.locals.nonce) {
          return "'strict-dynamic'";
        }
        return null;
      },
      "'unsafe-inline'",
      "'unsafe-eval'",
      'https:',
      "'report-sample'",
    ],
    'object-src': ["'none'"],
    'base-uri': ["'self'"],
    'default-src': ["'https:'"],
    'report-uri': [
      function reportUri() {
        let flowName = generateFlowName();
        flowName = flowName !== '' ? `?flow_name=${flowName}` : '';
        return `https://www.mercadopago.com.ar/security/csp-report${flowName}`;
      },
    ],
  },
  useDefaults: false,
  reportOnly: false,
};
