// cbt-web-navigation/src/navigation/header/components/user-menu/logged-in/options/index.js
const React = require('react');
const PropTypes = require('prop-types');
const SiteSwitcherOptions = require('../site-switcher');
const siteSwitcherProps = require('../site-switcher/props');
const types = require('../../../../../types');

const nameSpace = 'nav-header-cbt';

const LoggedInOptions = (props) => {
  const { myAccountText, logoutUrl, logoutText, showSiteSwitch } = props;

  return (
    <div id={`${nameSpace}__logged-options`}>
      <span className={`${nameSpace}__tail`} />
      <div className={`${nameSpace}__options`}>
        <div className={`${nameSpace}__option ${nameSpace}__my-account`}>
          <div className={`${nameSpace}__bold-text`}>{myAccountText}</div>
          <a id={`${nameSpace}__logout`} className={`${nameSpace}__blue-link`} href={logoutUrl}>{logoutText}</a>
        </div>
        {
          showSiteSwitch
          && <SiteSwitcherOptions {...siteSwitcherProps(props)} />
        }
      </div>
    </div>
  );
};

LoggedInOptions.defaultProps = {
  cbtInfo: null,
  logisticType: '',
  sites: null,
  logisticTypes: null,
};

LoggedInOptions.propTypes = {
  myAccountText: PropTypes.string.isRequired,
  showSiteSwitch: PropTypes.bool.isRequired,
  logoutUrl: PropTypes.string.isRequired,
  logoutText: PropTypes.string.isRequired,
  cbtInfo: PropTypes.shape(types.cbtInfoType),
  logisticType: PropTypes.string,
  sites: PropTypes.objectOf(PropTypes.shape(types.siteType)),
  logisticTypes: PropTypes.shape(types.logisticType),
};

module.exports = LoggedInOptions;
