// pi-web-navigation/src/navigation/widgets/user-menu/props.js
const userPictures = require('@frontend-navigations/commons/user-pictures');
const { getValue } = require('pi-navigation-config');
const cropUsername = require('../../utils/crop-username');

const userWidgetProps = (props) => {
  const {
    cdn,
    navigationUrls,
    urls,
    user,
    platform,
    navigationPIIData,
    browser,
    req,
  } = props;

  const userMenuVersion = getValue({
    key: 'userMenuVersion',
    platform,
    req,
  }) || '2.0.4';

  return {
    version: userMenuVersion,
    userId: user.id,
    platformId: platform.id,
    siteId: platform.siteId,
    domain: urls.mainDomain.link,
    baseUrl: cdn.link,
    email: navigationPIIData.email,
    show: props.device.desktop || false,
    title: `¡Hola, ${cropUsername(navigationPIIData.preferredName || user.nickname)}!`,
    userPicture: userPictures.getUserPicture(user, 'large', { browser }),
    myAccount: navigationUrls.myAccount,
    myQuotations: urls.myQuotations,
    history: urls.history,
    logout: navigationUrls.logout,
  };
};

module.exports = userWidgetProps;
