// ml-web-navigation/src/navigation/widgets/sign-up-message/index.js
const React = require('react');
const PropTypes = require('prop-types');
const Script = require('frontend-script');

const SignUpMessage = (props) => {
  const {
    baseUrl,
    version,
    show,
    isUserCompany,
    userName,
    userEmail,
    userNickname,
    changeNicknameUrl,
    companyFAQUrl,
    siteId,
  } = props;

  return show && (
    <Script src={`${baseUrl}/ui/navigation/${version}/scripts/messages-plus.js`}>
      {`
      if (window.freya) {
        window.freya.messages.showMessageForUser({
          user: {
            name: "${escape(userName)}",
            email: "${escape(userEmail)}",
            nickname: "${escape(userNickname)}",
            company: ${isUserCompany}
          },
          urls: {
            modifyUserUrl: "${changeNicknameUrl}",
            docsUrl: "${companyFAQUrl}",
          },
          siteId: "${siteId}"
        })
      }
      `}
    </Script>
  );
};

SignUpMessage.propTypes = {
  baseUrl: PropTypes.string.isRequired,
  version: PropTypes.string,
  show: PropTypes.bool.isRequired,
  isUserCompany: PropTypes.bool,
  userName: PropTypes.string,
  userEmail: PropTypes.string,
  userNickname: PropTypes.string,
  changeNicknameUrl: PropTypes.string,
  companyFAQUrl: PropTypes.string,
  siteId: PropTypes.string,
};

SignUpMessage.defaultProps = {
  isUserCompany: false,
  userName: '',
  userEmail: '',
  userNickname: '',
  changeNicknameUrl: '',
  companyFAQUrl: '',
  version: '4.4.1',
  siteId: null,
};

module.exports = SignUpMessage;
