// vis-web-navigation/src/navigation/footer/components/user-menu/logged-in/index.js
const React = require('react');
const PropTypes = require('prop-types');
const Image = require('frontend-image');

const namespace = 'nav-footer-user';
const FooterLoggedInUserMenu = (props) => {
  const {
    lowDensityPictureUrl,
    highDensityPictureUrl,
    hasPicture,
    lazyLoad,
    username,
    link,
  } = props;

  return (
    <p className={`${namespace} logged`}>
      <a href={link}>
        { hasPicture
          ? (<Image
              src={lowDensityPictureUrl}
              width="25"
              height="25"
              className="nav-footer-avatar-user-img"
              srcSet={`${highDensityPictureUrl} 2x`}
              lazyload={lazyLoad}
              onload="off"
          />) : <i className="nav-icon-user nav-footer-avatar-user-img" />}
        <strong>{username}</strong>
      </a>
    </p>
  );
};

FooterLoggedInUserMenu.propTypes = {
  hasPicture: PropTypes.bool,
  lowDensityPictureUrl: PropTypes.string,
  highDensityPictureUrl: PropTypes.string,
  lazyLoad: PropTypes.string,
  link: PropTypes.string.isRequired,
  username: PropTypes.string.isRequired,
};

FooterLoggedInUserMenu.defaultProps = {
  hasPicture: false,
  lowDensityPictureUrl: null,
  highDensityPictureUrl: null,
  lazyLoad: 'off',
};

module.exports = FooterLoggedInUserMenu;
