// ml-web-navigation/src/navigation/header/components/cart/index.js
const React = require('react');
const PropTypes = require('prop-types');
const { Cart } = require('nav-cart');
const tabIndexes = require('../../../utils/tab-index');

const CartComponent = (props) => {
  const { cartInfo, url, title, isMobile, cdnDomain, show, locale } = props;

  return (
    show && (
      <Cart
        data={cartInfo}
        url={url}
        title={title}
        isMobile={isMobile}
        cdnDomain={cdnDomain}
        tabIndex={tabIndexes('cart')}
        locale={locale}
      />
    )
  );
};

CartComponent.propTypes = {
  cartInfo: PropTypes.shape({
    badge_icon: PropTypes.object,
    location: PropTypes.object,
    carts: PropTypes.array,
  }),
  url: PropTypes.string.isRequired,
  title: PropTypes.string.isRequired,
  cdnDomain: PropTypes.string.isRequired,
  isMobile: PropTypes.bool.isRequired,
  show: PropTypes.bool.isRequired,
  locale: PropTypes.string,
};

CartComponent.defaultProps = {
  cartInfo: { badge_icon: null },
  locale: 'es',
};

module.exports = CartComponent;
