// ms-web-navigation/src/navigation/css/index.js
const React = require('react');
const propTypes = require('prop-types');
const CommonsCss = require('@frontend-navigations/commons/lib/css-component');
const {
  PLUS,
  INTERNAL,
} = require('../../constants/navigation-types');
const { isAnyPlus } = require('../../utils/navigation-types');

const {
  NAVIGATION_INTERNAL,
  NAVIGATION,
  NAVIGATION_LARGE,
  NAVIGATION_PLUS,
} = require('../../constants/css-files');

const Css = ({
  browser,
  navigationType,
  firstRequestInlineStyles,
  platformId,
  siteId,
  embedCss,
  feConfigEnvironment,
  cdn,
}) => {
  const loadLargeCss = browser && browser.name === 'ie' && parseInt(browser.major, 10) < 9;

  // navigation-internal__small.css
  const baseFiles = [(navigationType === INTERNAL && NAVIGATION_INTERNAL)
    // navigation-plus.css
    || (isAnyPlus(navigationType) && NAVIGATION_PLUS)
    // navigation.css
    || NAVIGATION];

  const files = baseFiles
    .concat(loadLargeCss ? NAVIGATION_LARGE : []) // navigation__large.css
    .map(name => ({
      name,
      firstRequestInlineStyles,
    }));

  return (
    <CommonsCss
      files={files}
      platformId={platformId}
      siteId={siteId}
      embedCss={embedCss}
      feConfigEnvironment={feConfigEnvironment}
      navigationType={navigationType}
      cdn={cdn}
      appName="ms-web-navigation"
    />
  );
};

Css.propTypes = {
  firstRequestInlineStyles: propTypes.bool,
  navigationType: propTypes.string,
  browser: propTypes.shape({
    name: propTypes.string,
    major: propTypes.number,
  }),
  platformId: propTypes.string,
  siteId: propTypes.string.isRequired,
  embedCss: propTypes.bool,
  feConfigEnvironment: propTypes.string,
  cdn: propTypes.shape({
    link: propTypes.string,
  }),
};

Css.defaultProps = {
  firstRequestInlineStyles: false,
  navigationType: PLUS,
  browser: null,
  platformId: 'MS',
  embedCss: false,
  feConfigEnvironment: null,
  cdn: null,
};

module.exports = Css;
