// mp-web-navigation/src/navigation/header/components/user-menu/navigation/landing-navigation/list-content/list-body/index.js
const React = require('react');
const PropTypes = require('prop-types');
const ListItem = require('../list-item');
const CollapsibleItem = require('../collapsible-item');

/* eslint-disable react/no-array-index-key */
const ListBody = ({ id, listItems, texts }) => (
  <>
    <input
      className="ln-collapsible-item__input"
      type="checkbox"
      id={`ln-collapsible-item-${id}`}
      value={`show_${id}`}
    />
    <ul className="ln-list-body">
      {listItems.map((item, index) => (
        <ListItem key={`listitem-key-${index}`} id={item.id} url={item.url} iconClass={item.icon} label={item.label} />
      ))}
      {(listItems.length >= 5) && (
        <CollapsibleItem id={id} labelState={{ open: texts.see_less, closed: texts.see_more }} />
      )}
    </ul>
  </>
);

ListBody.propTypes = {
  id: PropTypes.string.isRequired,
  listItems: PropTypes.arrayOf(
    PropTypes.shape({
      id: PropTypes.string,
      url: PropTypes.string,
      icon: PropTypes.string,
      label: PropTypes.string,
    }),
  ).isRequired,
  texts: PropTypes.shape({
    see_less: PropTypes.string,
    see_more: PropTypes.string,
  }).isRequired,
};

module.exports = ListBody;
