// mp-web-navigation/src/navigation/css/index.js
const React = require('react');
const propTypes = require('prop-types');
const CommonsCss = require('@frontend-navigations/commons/lib/css-component');
const {
  NAVIGATION_LARGE,
  NAVIGATION,
} = require('../../constants/css-files');
const { FULL } = require('../../constants/navigation-types');

const Css = ({
  browser,
  firstRequestInlineStyles,
  platformId,
  siteId,
  embedCss,
  feConfigEnvironment,
  navigationType,
  cdn,
  configGetter,
  configEnvironment,
  addConfigUpdateHandler,
}) => {
  const loadLargeCss = browser && browser.name === 'ie' && parseInt(browser.major, 10) < 9;

  const files = [NAVIGATION] // navigation.css
    .concat(loadLargeCss ? NAVIGATION_LARGE : []) // navigation__large.css
    .map(name => ({
      name,
      firstRequestInlineStyles,
    }));

  return (
    <CommonsCss
      files={files}
      platformId={platformId}
      siteId={siteId}
      embedCss={embedCss}
      feConfigEnvironment={feConfigEnvironment}
      navigationType={navigationType}
      cdn={cdn}
      configGetter={configGetter}
      configEnvironment={configEnvironment}
      addConfigUpdateHandler={addConfigUpdateHandler}
      appName="mp-web-navigation"
    />
  );
};

Css.propTypes = {
  firstRequestInlineStyles: propTypes.bool,
  browser: propTypes.shape({
    name: propTypes.string,
    major: propTypes.number,
  }),
  platformId: propTypes.string,
  siteId: propTypes.string,
  embedCss: propTypes.bool,
  feConfigEnvironment: propTypes.string,
  navigationType: propTypes.string,
  cdn: propTypes.shape({
    link: propTypes.string,
  }),
  configGetter: propTypes.func,
  configEnvironment: propTypes.string,
  addConfigUpdateHandler: propTypes.func,
};

Css.defaultProps = {
  firstRequestInlineStyles: false,
  browser: null,
  platformId: 'MP',
  siteId: 'MLA',
  embedCss: false,
  feConfigEnvironment: null,
  navigationType: FULL,
  cdn: null,
  configGetter: null,
  configEnvironment: null,
  addConfigUpdateHandler: null,
};

module.exports = Css;
