// cbt-web-navigation/src/constants/sites.js
module.exports = {
  MLA: {
    text: 'Argentina',
    countryId: 'AR',
    flag: 'https://http2.mlstatic.com/storage/cbt-statics/layout/header/flags/Argentina.svg',
  },
  MLM: {
    text: 'Mexico',
    countryId: 'MX',
    flag: 'https://http2.mlstatic.com/storage/cbt-statics/layout/header/flags/Mexico.svg',
  },
  MLB: {
    text: 'Brazil',
    countryId: 'BR',
    flag: 'https://http2.mlstatic.com/storage/cbt-statics/layout/header/flags/Brazil.svg',
  },
  MCO: {
    text: 'Colombia',
    countryId: 'CO',
    flag: 'https://http2.mlstatic.com/storage/cbt-statics/layout/header/flags/Colombia.svg',
  },
  MLC: {
    text: 'Chile',
    countryId: 'CL',
    flag: 'https://http2.mlstatic.com/storage/cbt-statics/layout/header/flags/Chile.svg',
  },
  MLV: {
    text: 'Venezuela',
    countryId: 'VE',
    flag: 'https://http2.mlstatic.com/storage/cbt-statics/layout/header/flags/Venezuela.svg',
  },
  MSV: {
    text: 'El Salvador',
    countryId: 'SV',
    flag: 'https://http2.mlstatic.com/storage/cbt-statics/layout/header/flags/Salvador.svg',
  },
  MNI: {
    text: 'Nicaragua',
    countryId: 'NI',
    flag: 'https://http2.mlstatic.com/storage/cbt-statics/layout/header/flags/Nicaragua.svg',
  },
  MGT: {
    text: 'Guatemala',
    countryId: 'GT',
    flag: 'https://http2.mlstatic.com/storage/cbt-statics/layout/header/flags/Guatemala.svg',
  },
  MEC: {
    text: 'Ecuador',
    countryId: 'EC',
    flag: 'https://http2.mlstatic.com/storage/cbt-statics/layout/header/flags/Ecuador.svg',
  },
  MHN: {
    text: 'Honduras',
    countryId: 'HN',
    flag: 'https://http2.mlstatic.com/storage/cbt-statics/layout/header/flags/Honduras.svg',
  },
  MPE: {
    text: 'Peru',
    countryId: 'PE',
    flag: 'https://http2.mlstatic.com/storage/cbt-statics/layout/header/flags/Peru.svg',
  },
  MLU: {
    text: 'Uruguay',
    countryId: 'UY',
    flag: 'https://http2.mlstatic.com/storage/cbt-statics/layout/header/flags/Uruguay.svg',
  },
  MRD: {
    text: 'República Dominicana',
    countryId: 'DO',
    flag: 'https://http2.mlstatic.com/storage/cbt-statics/layout/header/flags/Dominicana.svg',
  },
  MCR: {
    text: 'Costa Rica',
    countryId: 'CR',
    flag: 'https://http2.mlstatic.com/storage/cbt-statics/layout/header/flags/CostaRica.svg',
  },
  MBO: {
    text: 'Bolivia',
    countryId: 'BO',
    flag: 'https://http2.mlstatic.com/storage/cbt-statics/layout/header/flags/Bolivia.svg',
  },
  MPA: {
    text: 'Panamá',
    countryId: 'PA',
    flag: 'https://http2.mlstatic.com/storage/cbt-statics/layout/header/flags/Panama.svg',
  },
  MPY: {
    text: 'Paraguay',
    countryId: 'PY',
    flag: 'https://http2.mlstatic.com/storage/cbt-statics/layout/header/flags/Paraguay.svg',
  },
};
