// ml-web-navigation/src/navigation/header/components/cp/cp.desktop.js
const React = require('react');
const PropTypes = require('prop-types');
const ModalWidget = require('../../../widgets/modal');
const OnboardingWidget = require('../../../widgets/onboarding-cp');
const tabIndexes = require('../../../utils/tab-index');

const CPDesktop = ({ linkCP, label, title, modalWidget, onboardingWidget, showOnboarding }) => (
  <>
    <a {...linkCP} data-modal-action tabIndex={tabIndexes('cp')}>
      { title && <span className="nav-menu-cp-send">{title}</span> }
      <span className="nav-menu-link-cp"> {label}</span>
    </a>
    <ModalWidget {...modalWidget} />
    {showOnboarding && onboardingWidget && <OnboardingWidget {...onboardingWidget} />}
  </>
);

CPDesktop.propTypes = {
  linkCP: PropTypes.shape({
    className: PropTypes.string,
    href: PropTypes.string,
  }).isRequired,
  label: PropTypes.string.isRequired,
  title: PropTypes.string,
  modalWidget: PropTypes.shape({
    baseUrl: PropTypes.string,
    version: PropTypes.string,
  }).isRequired,
  onboardingWidget: PropTypes.object,
  showOnboarding: PropTypes.bool,
};

CPDesktop.defaultProps = {
  title: null,
  onboardingWidget: {},
  showOnboarding: true,
};

module.exports = CPDesktop;
