// ml-web-navigation/src/navigation/header/components/cp/props.js
const classNames = require('classnames');
const modalWidgetProps = require('../../../widgets/modal/props');
const onboardingWidgetProps = require('../../../widgets/onboarding-cp/props');
const desktop = require('../../../utils/is-desktop');

const cpProps = (props) => {
  const {
    cpInfo,
    navigationUrls,
    actualLocation,
    features,
    forceMobile = false,
  } = props;

  const encodedActualLocation = encodeURIComponent(actualLocation);
  const url = navigationUrls.cpLocation;
  const context = features.menuShippingLocation && features.menuShippingLocation.shippingMobileContext;
  const contextParam = context ? `&context=${context}` : '';
  const href = `${url.link}?go=${encodedActualLocation}${contextParam}`;
  const show = features.menuShippingLocation && features.menuShippingLocation.isShippingLocation;

  const isDesktop = desktop(props);
  const linkClassName = classNames({
    'nav-header-cp-anchor': !isDesktop || forceMobile,
    'nav-menu-cp': isDesktop && !forceMobile,
    'nav-menu-cp-logged': isDesktop && !forceMobile && cpInfo && cpInfo.title,
  });

  return {
    linkCP: {
      href,
      className: linkClassName,
      'data-js': 'cp', // If this ever changes, please update file navigation/client/cp-tracking.js
    },
    show: !!show,
    isDesktop,
    label: cpInfo && cpInfo.label ? cpInfo.label : url.name,
    title: cpInfo && cpInfo.title,
    modalWidget: modalWidgetProps(props),
    onboardingWidget: onboardingWidgetProps(props),
  };
};

module.exports = cpProps;
