// vis-web-navigation/src/navigation/header/components/user-menu/mobile/user-link/index.js
const React = require('react');
const PropTypes = require('prop-types');
const LinkTag = require('../../../../../common/link-tag');

const UserLink = ({ SideEffect, Aside, ...props }) => {
  const {
    link,
    icon,
    name,
    linkTag,
  } = props;

  return (
    <>
      <a
        {...link}
      >
        <i className={icon} />
        <span>
          {name}
          {Aside && <Aside {...props} />}
        </span>
        {linkTag && <LinkTag {...linkTag} />}
      </a>
      {SideEffect && <SideEffect {...props} />}
    </>
  );
};

UserLink.propTypes = {
  link: PropTypes.object.isRequired,
  icon: PropTypes.string.isRequired,
  name: PropTypes.string.isRequired,
  SideEffect: PropTypes.oneOfType([
    PropTypes.func,
    PropTypes.element,
  ]),
  Aside: PropTypes.oneOfType([
    PropTypes.func,
    PropTypes.element,
  ]),
  linkTag: PropTypes.object, // eslint-disable-line react/forbid-prop-types
};

UserLink.defaultProps = {
  SideEffect: null,
  Aside: null,
  linkTag: null,
};

module.exports = UserLink;
