// vis-web-navigation/src/navigation/header/components/nav-link/index.js
const React = require('react');
const PropTypes = require('prop-types');
const tabIndexes = require('../../../utils/tab-index');
const LinkTag = require('../../../common/link-tag');

const Link = (props) => {
  const { link, name, Component, SideEffect, type, linkTag, ...rest } = props;

  if (Component) return <Component {...rest} />;

  return (
    <>
      <a
        {...link}
        rel="nofollow"
        tabIndex={tabIndexes(type)}
      >
        {name}
        {linkTag && <LinkTag {...linkTag} small />}
      </a>
      {SideEffect && <SideEffect {...props} />}
    </>
  );
};

Link.propTypes = {
  link: PropTypes.shape({
    href: PropTypes.string,
    className: PropTypes.string,

  }).isRequired,
  name: PropTypes.oneOfType([
    PropTypes.string,
    PropTypes.func,
    PropTypes.element,
  ]).isRequired,
  Component: PropTypes.oneOfType([
    PropTypes.func,
    PropTypes.element,
  ]),
  SideEffect: PropTypes.oneOfType([
    PropTypes.func,
    PropTypes.element,
  ]),
  type: PropTypes.string,
  linkTag: PropTypes.object, // eslint-disable-line react/forbid-prop-types
};

Link.defaultProps = {
  Component: null,
  SideEffect: null,
  type: 'nav-link',
  linkTag: null,
};

module.exports = Link;
