// ml-web-navigation/src/navigation/widgets/onboarding-cp/index.js
const React = require('react');
const PropTypes = require('prop-types');
const Script = require('frontend-script');

const OnboardingCP = ({
  show,
  cpInfo,
  locale,
  domain,
  siteId,
  version,
  baseUrl,
}) => {
  const url = `${baseUrl}/frontend-assets/nav-widget-ml-onboarding-cp/v${version}/onboarding-cp.js`;
  return show && (
    <Script src={url}>{
      `(function () {
          new window.OnboardingCP({
            bus: window.meli,
            isNewBuyer: ${!cpInfo || !cpInfo.zipcode},
            isMobile: false,
            trigger: '.nav-menu-cp',
            locale: '${locale}',
            domain: '${domain}',
            siteId: '${siteId}',
            isInferred: ${!!(cpInfo && cpInfo.inferred)},
          });
        }());
    `
}
    </Script>
  );
};

OnboardingCP.propTypes = {
  show: PropTypes.bool.isRequired,
  baseUrl: PropTypes.string.isRequired,
  siteId: PropTypes.string.isRequired,
  domain: PropTypes.string.isRequired,
  locale: PropTypes.string.isRequired,
  cpInfo: PropTypes.object,
  version: PropTypes.string,
};

OnboardingCP.defaultProps = {
  version: '1.1.1',
  cpInfo: null,
};

module.exports = OnboardingCP;
