// cbt-web-navigation/src/navigation/header/components/user-menu/index.js
const React = require('react');
const PropTypes = require('prop-types');
const LoggedInMenu = require('./logged-in');
const GuestMenu = require('./guest-menu');
const loggedInMenuProps = require('./logged-in/props');
const guestMenuProps = require('./guest-menu/props');
const types = require('../../../types');

const nameSpace = 'nav-header-cbt';

const UserMenu = (props) => {
  const { isLoggedIn } = props;

  return (
    <div className={`${nameSpace}__menu`}>
      {
        isLoggedIn
          ? <LoggedInMenu {...loggedInMenuProps(props)} />
          : <GuestMenu {...guestMenuProps(props)} />
      }
    </div>
  );
};

UserMenu.defaultProps = {
  cbtInfo: null,
  user: null,
  sites: null,
  logisticTypes: null,
};

UserMenu.propTypes = {
  isLoggedIn: PropTypes.bool.isRequired,
  cbtInfo: PropTypes.shape(types.cbtInfoType),
  user: PropTypes.shape(types.userType),
  navigationPIIData: PropTypes.shape(types.navigationPIIData).isRequired,
  links: PropTypes.objectOf(PropTypes.shape(types.linkType)).isRequired,
  sites: PropTypes.objectOf(PropTypes.shape(types.siteType)),
  logisticTypes: PropTypes.shape(types.logisticType),
};

module.exports = UserMenu;
