// ml-web-navigation/src/navigation/footer/components/base-footer/index.js
/* eslint-disable react/forbid-prop-types */
const React = require('react');
const PropTypes = require('prop-types');
const Copyright = require('../copyright');
const DownloadApp = require('../../../widgets/download-app');
const HoneypathTrap = require('../honeypath-trap');

const namespace = 'nav-footer';
const BaseFooter = (props) => {
  const {
    copyright,
    children,
    banner,
    userMenuSection,
    footerExhibitor,
    tertiaries,
    baseFooter,
    downloadApp,
    honeypathTrap,
    isDesktop,
  } = props;

  return (
    <footer role="contentinfo" className={namespace}>
      { downloadApp && <DownloadApp position={downloadApp['footer-top']} /> }
      {banner}
      <div className={`${namespace}-user-info nav-bounds`}>
        {userMenuSection}
        <div className={`${namespace}-info-wrapper`}>
          <div className={`${namespace}-primaryinfo`}>
            {!isDesktop && children}
            <Copyright {...copyright} />
            {isDesktop && children}
          </div>
          { baseFooter.extraCompanyName && <p className="nav-footer-secondaryinfo">{baseFooter.extraCompanyName}</p> }
        </div>
        {footerExhibitor}
      </div>
      {tertiaries && (
        <div className="nav-footer-tertiary-info nav-bounds">
          {tertiaries}
        </div>
      )}
      { downloadApp && <DownloadApp position={downloadApp['footer-bottom']} /> }
      <HoneypathTrap {...honeypathTrap} />
    </footer>
  );
};

BaseFooter.propTypes = {
  copyright: PropTypes.object.isRequired,
  children: PropTypes.oneOfType([
    PropTypes.arrayOf(PropTypes.node),
    PropTypes.node,
  ]),
  banner: PropTypes.oneOfType([
    PropTypes.arrayOf(PropTypes.node),
    PropTypes.node,
  ]),
  footerExhibitor: PropTypes.oneOfType([
    PropTypes.arrayOf(PropTypes.node),
    PropTypes.node,
  ]),
  userMenuSection: PropTypes.oneOfType([
    PropTypes.arrayOf(PropTypes.node),
    PropTypes.node,
  ]),
  tertiaries: PropTypes.oneOfType([
    PropTypes.arrayOf(PropTypes.node),
    PropTypes.node,
  ]),
  downloadApp: PropTypes.object,
  honeypathTrap: PropTypes.shape({
    label: PropTypes.string,
    url: PropTypes.string,
    webView: PropTypes.bool,
    platform: PropTypes.string,
    siteId: PropTypes.string,
    deviceType: PropTypes.string,
  }),
};

BaseFooter.defaultProps = {
  children: null,
  banner: null,
  userMenuSection: null,
  footerExhibitor: null,
  tertiaries: null,
  honeypathTrap: null,
  downloadApp: null,
};

module.exports = BaseFooter;
