// frontend-analytics/config/optimize.js
const optimizeConfig = {
  ml: {
    mla: { id: 'GTM-KG9L85Q' },
    mlb: { id: 'GTM-NKDZGQK' },
    mlm: { id: 'GTM-MZDGLTZ' },
    mbo: { id: 'GTM-P6Q2T25' },
    mlc: { id: 'GTM-KVFTG7M' },
    mco: { id: 'GTM-W42JSM9' },
    mcr: { id: 'GTM-WLGTWJJ' },
    mcu: { id: 'GTM-5SP32QT' },
    mec: { id: 'GTM-MQGSN6J' },
    msv: { id: 'GTM-NJC6ZMH' },
    mgt: { id: 'GTM-PNNZCKV' },
    mhn: { id: 'GTM-ND9H7TD' },
    mni: { id: 'GTM-NZSHWWH' },
    mpa: { id: 'GTM-5BL2R4X' },
    mpy: { id: 'GTM-TPQKS8B' },
    mpe: { id: 'GTM-5KVCTNR' },
    mpt: { id: 'GTM-TJNPQ77' },
    mrd: { id: 'GTM-PQP3K9R' },
    mlu: { id: 'GTM-W2JN88W' },
    mlv: { id: 'GTM-PG5TPV5' },
  },
  mp: {
    mla: { id: 'GTM-56V8XPP' },
    mlb: { id: 'GTM-K4G4SRW' },
    mlm: { id: 'GTM-T5MSJRV' },
    mlc: { id: 'GTM-KSBHVX8' },
    mco: { id: 'GTM-P9XT2XN' },
    mpe: { id: 'GTM-K2SRH7N' },
    mlu: { id: 'GTM-K6VHKLK' },
    mlv: { id: 'GTM-WJ9MKBV' },
    mec: { id: 'OPT-KFS2V3X' },
  },
};

module.exports = optimizeConfig;
