// frontend-mshops-detection/src/config/default.js
module.exports = {
  ALLOWED_HEADERS_TO_BYPASS: [
    'X-Frontend-Wrapper-Scope',
    'x-robots',
  ],
  CLIENT: {
    FRONTEND_WRAPPER: {
      /**
       * Since our frontend WON'T work unless there's an answer from frontend-wrapper,
       * we can be a bit lenient in its configuration values.
       * Worst case scenario: 1000 + 1100 + 1300 = 3400ms to reply
       * TODO: Let each fury application to add its own configuration if needed
       */
      timeout: 1000, // The number of ms until the connection is aborted
      retry: false, // Disable any retries
    },
  },
  ENDPOINT: {
    GUEST_DETECTION: (domain, jwt) => `https://${domain}/mercadoshops/check-session?signed=${jwt}`,
    FRONTEND_WRAPPER: '/shops/frontend-wrapper',
  },
  ML_SITES: {
    MLA: { domain: 'www.mercadolibre.com.ar' },
    MLB: { domain: 'www.mercadolivre.com.br' },
    MLC: { domain: 'www.mercadolibre.cl' },
    MCO: { domain: 'www.mercadolibre.com.co' },
    MLM: { domain: 'www.mercadolibre.com.mx' },
  },
  COOKIES: {
    GUEST_CART_ID: '_ml_gc',
    CP: {
      name: 'cp',
      expMin: 30,
    },
    IS_GUEST: {
      name: 'mshops_isGuest',
      expMin: 30,
    },
  },
  features: {
    robots: false,
    fetchShopData: {
      enabled: true,
      devShopDomain: null,
    },
    guestDetection: {
      enabled: false,
      paths: [
        /^\//,
      ],
      deviceTypes: ['desktop'],
      jwtSecret: null,
      jwtTTL: '30s',
      queries: {
        isGuest: 'mshops-cookie-isguest',
        cp: 'mshops-cookie-cp',
        timestamp: 'mshops-redirection-timestamp',
      },
    },
  },
};
