// cbt-web-navigation/src/prepare-to-render/services/footer/index.js
// eslint-disable-next-line import/no-extraneous-dependencies
const feConfig = require('frontend-config');
const { getValue } = require('cbt-navigation-config');

const getFooterTexts = (platformId, siteId, feConfigEnvironment) => {
  const currentYear = new Date().getFullYear();
  const companyName = feConfig.get('companyName', platformId, siteId, feConfigEnvironment);
  return {
    copyright: `Copyright © ${currentYear} ${companyName}`,
  };
};

const execute = (data) => {
  const {
    platform,
    req,
    feConfigEnvironment,
    isDesktop,
  } = data;

  const cbtConfigUrls = getValue({
    key: 'url',
    platform,
    req,
  });
  const feConfigUrls = feConfig.get('url', platform.id, platform.siteId, feConfigEnvironment);

  const links = getValue({
    key: 'desktopFooterLinks',
    platform,
    req,
  }).map(link => cbtConfigUrls[link] || feConfigUrls[link]);

  return Promise.resolve({
    links,
    texts: getFooterTexts(platform.id, platform.siteId, feConfigEnvironment),
    platformId: platform.id,
    siteId: platform.siteId,
    isDesktop,
    feConfigEnvironment,
  });
};

const generateServiceContext = (req, res, props) => {
  const { platform, device } = req;

  return {
    platform,
    isDesktop: device.desktop,
    feConfigEnvironment: props.feConfigEnvironment,
    req,
  };
};

module.exports = (req, res, props) => execute(generateServiceContext(req, res, props));
