// cbt-web-navigation/src/navigation/header/components/user-menu/logged-in/site-switcher/props.js
const sitePriorities = {
  MLM: 1,
  MLB: 2,
  MLA: 3,
  MLC: 4,
  MCO: 5,
};

const logPriorities = {
  fulfillment: 1,
  '3pl': 2,
  remote: 3,
};

const sortByFactory = (prioritiesHash, property) => (a, b) => {
  const siteAPriority = prioritiesHash[a[property]]; // eslint-disable-line security/detect-object-injection
  const siteBPriority = prioritiesHash[b[property]]; // eslint-disable-line security/detect-object-injection

  if (siteAPriority) {
    if (siteBPriority) {
      return siteAPriority - siteBPriority;
    }
    return -1;
  } if (siteBPriority) {
    return 1;
  }

  // The ECMAscript standard does not guarantee the sort behavior
  return 0;
};

const sortByPriority = sortByFactory(sitePriorities, 'site_id');
const sortByLogPriority = sortByFactory(logPriorities, 'logistic_type');

/**
 * This function sorts the sites to show in the switcher by the priorities defined in above constants.
 * For example: 1- MLM Fulfillment, 2- MLM 3pl, 3- MLM remote, 4- MLB remote, 5- MLA remote
 * If a site is not listed, will be added to the bottom of the list.
 * @param a
 * @param b
 * @return {boolean}
 */
const sortSites = (a, b) => sortByPriority(a, b) || sortByLogPriority(a, b);

const siteSwitcherProps = (props) => {
  const { cbtInfo, logisticType, sites, logisticTypes } = props;
  const { siteUsers, currentSiteId } = cbtInfo;
  siteUsers.sort(sortSites);

  return {
    currentSiteId,
    logisticType,
    siteUsers,
    sites,
    logisticTypes,
  };
};

module.exports = siteSwitcherProps;
