// ml-web-navigation/src/navigation/widgets/loan-message/i18n.js
/* eslint-disable max-len */
// TODO: This should be replaced by getting loan-message translations from an api call.
const i18n = {
  no_charge_period: {
    es: {
      singular: 'Tienes 1 cuota de Mercado Crédito vencida. Págala ahora con tarjeta de débito o efectivo para evitar recargos.',
      plural: 'Tienes {N} cuotas de Mercado Crédito vencidas. Págalas ahora con tarjeta de débito o efectivo para evitar recargos.',
    },
    es_AR: {
      singular: 'Tenés 1 cuota de Mercado Crédito vencida. Pagala ahora con tarjeta de débito o efectivo para evitar recargos.',
      plural: 'Tenés {N} cuotas de Mercado Crédito vencidas. Pagalas ahora con tarjeta de débito o efectivo para evitar recargos.',
    },
    es_MX: {
      singular: 'Tienes 1 mensualidad de Mercado Crédito vencida. Págala ahora con tarjeta de débito o efectivo para evitar recargos.',
      plural: 'Tienes {N} mensualidades de Mercado Crédito vencidas. Págalas ahora con tarjeta de débito o efectivo para evitar recargos.',
    },
    pt: {
      singular: 'Você possui 1 parcela do Mercado Crédito vencida. Pague-a agora com boleto para evitar custos adicionais.',
      plural: 'Você possui {N} parcelas do Mercado Crédito vencidas. Pague-as agora com boleto para evitar custos adicionais.',
    },
  },
  fixed_charge_period_1: {
    es: {
      singular: 'Tu cuota vencida de Mercado Crédito tiene un recargo. Págala para volver a comprar con Mercado Crédito.',
      plural: 'Tus cuotas vencidas de Mercado Crédito tienen recargos. Págalas para volver a comprar con Mercado Crédito.',
    },
    es_AR: {
      singular: 'Tu cuota vencida de Mercado Crédito tiene un recargo. Pagala para volver a comprar con Mercado Crédito.',
      plural: 'Tus cuotas vencidas de Mercado Crédito tienen recargos. Pagalas para volver a comprar con Mercado Crédito.',
    },
    es_MX: {
      singular: 'Tu mensualidad vencida de Mercado Crédito tiene un recargo. Págala para volver a comprar con Mercado Crédito.',
      plural: 'Tus mensualidades vencidas de Mercado Crédito tienen recargos. Págalas para volver a comprar con Mercado Crédito.',
    },
    pt: {
      singular: 'Sua parcela vencida do Mercado Crédito tem custos adicionais. Pague-a agora para voltar a comprar com o Mercado Crédito.',
      plural: 'Suas parcelas vencidas do Mercado Crédito têm custos adicionais. Pague-as agora para voltar a comprar com o Mercado Crédito.',
    },
  },
  fixed_charge_period_2: {
    es: {
      singular: 'Tienes 1 cuota de Mercado Crédito vencida con recargos. Paga tu deuda ahora y evita recargos diarios.',
      plural: 'Tienes {N} cuotas de Mercado Crédito vencidas con recargos. Paga tu deuda ahora y evita recargos diarios.',
    },
    es_AR: {
      singular: 'Tenés 1 cuota de Mercado Crédito vencida con recargos. Pagá tu deuda ahora y evitá recargos diarios.',
      plural: 'Tenés {N} cuotas de Mercado Crédito vencidas con recargos. Pagá tu deuda ahora y evitá recargos diarios.',
    },
    es_MX: {
      singular: 'Tienes 1 mensualidad de Mercado Crédito vencida con recargos. Paga tu deuda ahora y evita recargos diarios.',
      plural: 'Tienes {N} mensualidades de Mercado Crédito vencidas con recargos. Paga tu deuda ahora y evita recargos diarios.',
    },
    pt: {
      singular: 'Você possui 1 parcela do Mercado Crédito vencida com custos adicionais. Pague a sua dívida agora mesmo para evitar custos adicionais diários.',
      plural: 'Você possui {N} parcelas do Mercado Crédito vencidas com custos adicionais. Pague a sua dívida agora mesmo para evitar custos adicionais diários.',
    },
  },
  daily_charge_period: {
    es: {
      singular: 'Tu deuda con Mercado Crédito aumentará día a día. Págala ahora y evita que se incremente aún más con el próximo vencimiento.',
      plural: 'Tu deuda con Mercado Crédito aumentará día a día. Págala ahora y evita que se incremente aún más con el próximo vencimiento.',
    },
    es_AR: {
      singular: 'Tu deuda con Mercado Crédito aumentará día a día. Pagala ahora y evitá que se incremente aún más con el próximo vencimiento.',
      plural: 'Tu deuda con Mercado Crédito aumentará día a día. Pagala ahora y evitá que se incremente aún más con el próximo vencimiento.',
    },
    es_MX: {
      singular: 'Tu deuda con Mercado Crédito aumentará día a día. Págala ahora y evita que se incremente aún más con la próxima fecha de pago.',
      plural: 'Tu deuda con Mercado Crédito aumentará día a día. Págala ahora y evita que se incremente aún más con la próxima fecha de pago.',
    },
    pt: {
      singular: 'Sua dívida com o Mercado Crédito aumentará diariamente. Pague-a agora mesmo para evitar que ela fique ainda mais alta com o próximo vencimento.',
      plural: 'Sua dívida com o Mercado Crédito aumentará diariamente. Pague-a agora mesmo para evitar que ela fique ainda mais alta com o próximo vencimento.',
    },
  },
};

module.exports = i18n;
