// cbt-web-navigation/rollup-css.config.js
import scss from 'rollup-plugin-scss';
import { version } from './package.json';
import {
  join,
  resolve,
} from 'path';
import postcss from 'postcss';
import autoprefixer from 'autoprefixer';
import url from 'postcss-url';
import postCssBanner from 'postcss-banner';
import normalizeCharset from 'postcss-normalize-charset';
import {
  writeFileSync,
  readdirSync,
} from 'fs';
import copy from 'rollup-plugin-copy';

const isDevelopment = process.env.NODE_ENV === 'development';

const businessUnit = 'mercadolibre';
const outputFolder = join(resolve('.'), `object-storage-assets/ui-navigation/${version}/${businessUnit}`);

const cdnBase = isDevelopment
  ? `http://localhost:8081`
  : `https://http2.mlstatic.com`;
const urlBase = `${cdnBase}/frontend-assets/cbt-web-navigation/ui-navigation/${version}`;

const businessUnitPath = 'business-unit-path/';
const entryPointsRelativeFolder = './src/css/entry-points';
const entryPointsAbsoluteFolder = join(resolve('.'), entryPointsRelativeFolder);

const bannerContent = `Navigation
@platform "${businessUnit}"
@version ${version}
@author MercadoLibre.com`;

const getFileConfig = (name, copyImageFiles) => {
  const jsFileName = `${entryPointsRelativeFolder}/${name}.js`;

  writeFileSync(join(entryPointsAbsoluteFolder, `${name}.js`), `import './${name}.scss';`);

  return {
    input: jsFileName,
    output: {
      file: join(outputFolder, `${name}.js`),
      format: 'esm',
    },
    plugins: [
      scss({
        processor: () => postcss([
          autoprefixer({
            overrideBrowserslist: [
              'chrome >= 22',
              'firefox >= 24',
              'android >= 4.4',
              'safari >= 7',
              'ie 11',
              'not dead',
            ],
            flexbox: 'no-2009',
          }),
          url([{
            filter: `${businessUnitPath}*`,
            url: asset => `${urlBase}/${businessUnit}/cbt/${asset.url.slice(businessUnitPath.length)}`,
          }]),
          postCssBanner({
            banner: bannerContent,
          }),
          normalizeCharset(true),
        ]),
        outputStyle: 'compressed',
      }),
    ].concat(copyImageFiles
      ? copy({
        targets: [
          { src: 'src/images/*', dest: join(outputFolder, 'cbt') },
        ]
      })
      : []
    ),
  };
};

const fileExtension = '.scss';
const entryPoints = readdirSync(entryPointsAbsoluteFolder, { withFileTypes: true })
  .filter((entry) => !entry.isDirectory() && entry.name.endsWith(fileExtension))
  .map(({ name }, index) => getFileConfig(name.slice(0, - fileExtension.length), !index));

if (!entryPoints.length) {
  throw new Error('No scss files where found');
}

export default entryPoints;
