// ragnar/lib/services/suspicious-traffic.js
/**
 * Modules dependencies
 */
const restclient = require('frontend-restclient')({
  timeout: 300,
});

const DEFAULT_SECURITY_LEVEL = 'low';

/**
 * Service interface
 */
class SuspiciousTrafficService {
  static getClientStatus({
    siteId, ip, host, deviceType, userAgent, forwardedFor, accept, context,
  }) {
    return restclient.post('/auth/traffic_policies/site_traffic/state', {
      context,
      data: {
        http_request: {
          ip,
          headers: {
            Accept: accept,
            'X-Device-Type': deviceType.toUpperCase(),
            'User-Agent': userAgent,
            'X-Forwarded-For': forwardedFor,
            Host: host,
          },
        },
        site_id: siteId,
        context_data: {},
      },
    })
      .then((res) => {
        if (res.data && res.data.result) {
          return res.data.result;
        }

        return DEFAULT_SECURITY_LEVEL;
      })
      .catch(() => DEFAULT_SECURITY_LEVEL);
  }
}

/**
 * Expose Service
 */
module.exports = SuspiciousTrafficService;
