// @frontend-performance/image-lazy-loading/tests/utils.js
/* eslint-disable no-unused-expressions */
const defaults = {
  lazyLoadClassName: 'lazy-loadable',
  onloadClassName: 'onload-loadable',
  src: 'src',
  dataSrc: 'data-src',
  dataSrcFallback: 'data-src-fallback',
  srcSet: 'srcset',
  dataSrcSet: 'data-srcset',
  loadModeLazy: 'lazy',
  loadModeOnload: 'onload',
};

const appendImage = (options) => {
  const { classes, src, dataSrc, srcSet, dataSrcSet, dataSrcFallback } = options;
  const image = document.createElement('img');

  for (let i = 0; i < classes.length; i += 1) {
    const clazz = classes[i];
    image.classList.add(clazz);
  }

  dataSrc && image.setAttribute(defaults.dataSrc, dataSrc);
  src && image.setAttribute(defaults.src, src);
  srcSet && image.setAttribute(defaults.srcSet, srcSet);
  dataSrcSet && image.setAttribute(defaults.dataSrcSet, dataSrcSet);
  dataSrcFallback && image.setAttribute(defaults.dataSrcFallback, dataSrcFallback);

  document.body.appendChild(image);

  return image;
};

const triggerLoadEvent = () => {
  const evt = document.createEvent('Event');
  evt.initEvent('load', false, false);
  window.dispatchEvent(evt);
};

const triggerErrorEvent = (element) => {
  const evt = document.createEvent('Event');
  evt.initEvent('error', false, false);
  element.dispatchEvent(evt);
};

module.exports = {
  appendImage,
  defaults,
  triggerLoadEvent,
  triggerErrorEvent,
};
