// frontend-logger/src/Logger.js
/**
 * Module dependencies
 */
const {
  LOG_LEVEL,
  DEVELOPMENT,
  TEST,
} = require('frontend-env');
const { createLogger, format, transports } = require('winston');
const json2tags = require('./json2tags');

/**
 * Available logger methods
 */
const methods = [
  'info', 'warn', 'error',
  'verbose', 'debug', 'silly',
];

/**
 * Custom format
 */
const defaultMessageFormat = format.printf(({ timestamp, level, message }) => `${timestamp} - ${level}: ${message}`);

const devModeformat = format.combine(
  format.timestamp(),
  defaultMessageFormat,
  format.colorize({ all: true }),
);

const prodModeFormat = format.combine(
  format.timestamp(),
  defaultMessageFormat,
);

/**
 * Winston initialization
 */
const winstonInit = createLogger({
  exitOnError: false,
  transports: [
    new transports.Console({
      level: methods.includes(LOG_LEVEL) ? LOG_LEVEL : 'info',
      stderrLevels: ['error'],
      handleExceptions: true,
      format: DEVELOPMENT || TEST ? devModeformat : prodModeFormat,
    }),
  ],
});

/**
 * Logger constructor
 */
function Logger(name, options = {}) {
  this.name = name;
  this.options = options;
  this.client = winstonInit;

  return this;
}

/**
 * Add available method to the prototype
 */
methods.forEach((method) => {
  // eslint-disable-next-line func-names
  Logger.prototype[method] = function (msg, tags = {}) {
    if (this.name) {
      tags.name = this.name;
    }
    tags.level = method.toUpperCase();
    this.client[method](`${msg} - ${json2tags(tags)}`);
  };
});

/**
 * Expose Logger
 */
module.exports = Logger;
