// ml-web-navigation/src/navigation/header/plus/index.js
const React = require('react');
const PropTypes = require('prop-types');
const Header = require('../components/base-header');
const { types } = require('../../../layout-type-detection');
const Searchbox = require('../components/searchbox');
const Cart = require('../components/cart');
const CP = require('../components/cp');
const NavMenu = require('../components/nav-menu');
const Exhibitor = require('../components/exhibitor');

const HeaderPlus = ({ searchbox, ...props }) => {
  const plusType = types.PLUS;
  const { cart, cp, navMenu, exhibitor, isDesktop } = props;

  return (
    <Header
      type={plusType}
      {...props}
    >
      <Searchbox {...searchbox} />
      <NavMenu {...navMenu} />
      <Cart {...cart} />
      { isDesktop && <Exhibitor {...exhibitor} /> }
      <CP {...cp} showOnboarding={false} />
    </Header>
  );
};

HeaderPlus.propTypes = {
  searchbox: PropTypes.object.isRequired,
  navMenu: PropTypes.object.isRequired,
  cart: PropTypes.object.isRequired,
  cp: PropTypes.object.isRequired,
  isDesktop: PropTypes.bool.isRequired,
  exhibitor: PropTypes.object,
};

HeaderPlus.defaultProps = {
  exhibitor: {},
};

module.exports = HeaderPlus;
