// ragnar/lib/nordicPage/middleware/nordic-pages-property.js
const { getPageResources } = require('../utils');
const { getMiddlewarePropsIfApply } = require('../page-segments');


const defineNordicPagePropAtReq = (req, { paths, ...options }) => {
  const defaults = {
    pageResources: getPageResources(paths, req.device),
    serverSidePropsData: {},
    pageSettingsData: {},
  };

  const finalProps = { ...defaults, ...options };

  Object.defineProperty(req, 'nordicPages', { value: finalProps });
};

const definePropertyFor404HttpError = (paths) => function addNordicPagesProperty(req, res, next) {
  const options = {
    paths,
    statusCode: 404,
  };

  defineNordicPagePropAtReq(req, options);

  next();
};

const definePropertyForNavigableNordicPage = ({ paths, route, pageHooks }) => function addNordicPagesProperty(req, res, next) {
  const options = {
    paths,
    route,
    pageHooks,
  };

  defineNordicPagePropAtReq(req, options);

  next();
};

const initializeServerSidePropsData = (req) => {
  req.nordicPages.serverSidePropsData = { settings: {}, pageState: {}, done: null };
};

const addServerSidePropsData = ({ req, pageState, settings }) => {
  const { view } = req.nordicPages.pageResources;
  const serverSideProps = getMiddlewarePropsIfApply({
    name: 'GET_SERVER_SIDE_PROPS',
    props: {
      pageState,
      settings,
    },
    view,
  });

  req.nordicPages.serverSidePropsData = serverSideProps;
};

const initializePageSettingsData = (req, defaultPageSettings) => {
  req.nordicPages.pageSettingsData = {
    pageSettings: defaultPageSettings,
    layoutOptions: null,
    navigationOptions: null,
  };
};

const addPageSettingsData = ({ req, pageSettings, layoutOptions, navigationOptions }) => {
  const { view } = req.nordicPages.pageResources;
  const pageSettingsData = getMiddlewarePropsIfApply({
    name: 'SET_PAGE_SETTINGS',
    props: {
      pageSettings,
      layoutOptions,
      navigationOptions,
    },
    view,
  });

  req.nordicPages.pageSettingsData = pageSettingsData;
};

const addDoneToServerSidePropsData = (req, done) => {
  req.nordicPages.serverSidePropsData.done = (done || null);
};

const NordicPagesProperty = {
  definePropertyFor404HttpError,
  definePropertyForNavigableNordicPage,
  initializeServerSidePropsData,
  addServerSidePropsData,
  initializePageSettingsData,
  addPageSettingsData,
  addDoneToServerSidePropsData,
};

module.exports = NordicPagesProperty;
