// frontend-layout/config/platforms/ml.js
module.exports = {
  MLA: {
    i18n: {
      favorites: 'Favoritos',
      signinAs: 'Ingresaste como',
      notifications: 'Notificaciones',
      noNotifications: 'Por ahora, no hay nada acá.',
      myAccountDashboard: 'Resumen',
      searchIconLabel: 'Buscar',
      searchInputLabel: 'Ingresá lo que quieras encontrar',
      searchPlaceholder: 'Buscar productos, marcas y más…',
      searchPlaceholderMobile: 'Estoy buscando…',
    },
    footer: {
      items: ['jobs', 'tyc', 'politics', 'helpMain'],
    },
    styles: {
      useProximaNova: true,
    },
  },
  MLB: {
    i18n: {
      favorites: 'Favoritos',
      signinAs: 'Assinado como',
      notifications: 'Avisos',
      noNotifications: ' Por enquanto, não há nada aqui',
      myAccountDashboard: 'Resumo',
      searchIconLabel: 'Buscar',
      searchInputLabel: 'Digite o que você quer encontrar',
      searchPlaceholder: 'Buscar produtos, marcas e muito mais…',
      searchPlaceholderMobile: 'Estou buscando…',
    },
    footer: {
      items: ['jobs', 'tyc', 'politics', 'helpMain'],
    },
    styles: {
      useProximaNova: true,
    },
  },
  MLC: {
    i18n: {
      favorites: 'Favoritos',
      signinAs: 'Ingresaste como',
      notifications: 'Notificaciones',
      noNotifications: 'Nada por acá',
      myAccountDashboard: 'Resumen',
      searchIconLabel: 'Buscar',
      searchInputLabel: 'Ingresa lo que quieras encontrar',
      searchPlaceholder: 'Buscar productos, marcas y más…',
      searchPlaceholderMobile: 'Estoy buscando…',
    },
    footer: {
      items: ['jobs', 'tyc', 'politics', 'helpMain'],
    },
    styles: {
      useProximaNova: true,
    },
  },
  MLM: {
    i18n: {
      favorites: 'Favoritos',
      signinAs: 'Ingresaste como',
      notifications: 'Notificaciones',
      noNotifications: 'Nada por acá',
      myAccountDashboard: 'Resumen',
      searchIconLabel: 'Buscar',
      searchInputLabel: 'Ingresa lo que quieras encontrar',
      searchPlaceholder: 'Buscar productos, marcas y más…',
      searchPlaceholderMobile: 'Estoy buscando…',
    },
    footer: {
      items: ['jobs', 'tyc', 'politics', 'helpMain'],
    },
    styles: {
      useProximaNova: true,
    },
  },
  MLU: {
    i18n: {
      favorites: 'Favoritos',
      signinAs: 'Ingresaste como',
      notifications: 'Notificaciones',
      noNotifications: 'Nada por acá',
      myAccountDashboard: 'Resumen',
      searchIconLabel: 'Buscar',
      searchInputLabel: 'Ingresa lo que quieras encontrar',
      searchPlaceholder: 'Buscar productos, marcas y más…',
      searchPlaceholderMobile: 'Estoy buscando…',
    },
    footer: {
      items: ['jobs', 'tyc', 'politics', 'helpMain'],
    },
    styles: {
      useProximaNova: true,
    },
  },
  MLV: {
    i18n: {
      favorites: 'Favoritos',
      signinAs: 'Ingresaste como',
      notifications: 'Notificaciones',
      noNotifications: 'Nada por acá',
      myAccountDashboard: 'Resumen',
      searchIconLabel: 'Buscar',
      searchInputLabel: 'Ingresa lo que quieras encontrar',
      searchPlaceholder: 'Buscar productos, marcas y más…',
      searchPlaceholderMobile: 'Estoy buscando…',
    },
    footer: {
      items: ['jobs', 'tyc', 'politics', 'helpMain'],
    },
    styles: {
      useProximaNova: true,
    },
  },
  MBO: {
    i18n: {
      favorites: 'Favoritos',
      signinAs: 'Ingresaste como',
      notifications: 'Notificaciones',
      noNotifications: 'Nada por acá',
      myAccountDashboard: 'Resumen',
      searchIconLabel: 'Buscar',
      searchInputLabel: 'Ingresa lo que quieras encontrar',
      searchPlaceholder: 'Buscar productos, marcas y más…',
      searchPlaceholderMobile: 'Estoy buscando…',
    },
    footer: {
      items: ['tyc', 'politics', 'helpMain'],
    },
    styles: {
      useProximaNova: true,
    },
  },
  MCO: {
    i18n: {
      favorites: 'Favoritos',
      signinAs: 'Ingresaste como',
      notifications: 'Notificaciones',
      noNotifications: 'Nada por acá',
      myAccountDashboard: 'Resumen',
      searchIconLabel: 'Buscar',
      searchInputLabel: 'Ingresa lo que quieras encontrar',
      searchPlaceholder: 'Buscar productos, marcas y más…',
      searchPlaceholderMobile: 'Estoy buscando…',
    },
    footer: {
      items: ['jobs', 'tyc', 'politics', 'helpMain'],
    },
    styles: {
      useProximaNova: true,
    },
  },
  MCR: {
    i18n: {
      favorites: 'Favoritos',
      signinAs: 'Ingresaste como',
      notifications: 'Notificaciones',
      noNotifications: 'Nada por acá',
      myAccountDashboard: 'Resumen',
      searchIconLabel: 'Buscar',
      searchInputLabel: 'Ingresa lo que quieras encontrar',
      searchPlaceholder: 'Buscar productos, marcas y más…',
      searchPlaceholderMobile: 'Estoy buscando…',
    },
    footer: {
      items: ['tyc', 'politics', 'helpMain'],
    },
    styles: {
      useProximaNova: true,
    },
  },
  MRD: {
    i18n: {
      favorites: 'Favoritos',
      signinAs: 'Ingresaste como',
      notifications: 'Notificaciones',
      noNotifications: 'Nada por acá',
      myAccountDashboard: 'Resumen',
      searchIconLabel: 'Buscar',
      searchInputLabel: 'Ingresa lo que quieras encontrar',
      searchPlaceholder: 'Buscar productos, marcas y más…',
      searchPlaceholderMobile: 'Estoy buscando…',
    },
    footer: {
      items: ['tyc', 'politics', 'helpMain'],
    },
    styles: {
      useProximaNova: true,
    },
  },
  MEC: {
    i18n: {
      favorites: 'Favoritos',
      signinAs: 'Ingresaste como',
      notifications: 'Notificaciones',
      noNotifications: 'Nada por acá',
      myAccountDashboard: 'Resumen',
      searchIconLabel: 'Buscar',
      searchInputLabel: 'Ingresa lo que quieras encontrar',
      searchPlaceholder: 'Buscar productos, marcas y más…',
      searchPlaceholderMobile: 'Estoy buscando…',
    },
    footer: {
      items: ['tyc', 'politics', 'helpMain'],
    },
    styles: {
      useProximaNova: true,
    },
  },
  MGT: {
    i18n: {
      favorites: 'Favoritos',
      signinAs: 'Ingresaste como',
      notifications: 'Notificaciones',
      noNotifications: 'Nada por acá',
      myAccountDashboard: 'Resumen',
      searchIconLabel: 'Buscar',
      searchInputLabel: 'Ingresa lo que quieras encontrar',
      searchPlaceholder: 'Buscar productos, marcas y más…',
      searchPlaceholderMobile: 'Estoy buscando…',
    },
    footer: {
      items: ['tyc', 'politics', 'helpMain'],
    },
    styles: {
      useProximaNova: true,
    },
  },
  MHN: {
    i18n: {
      favorites: 'Favoritos',
      signinAs: 'Ingresaste como',
      notifications: 'Notificaciones',
      noNotifications: 'Nada por acá',
      myAccountDashboard: 'Resumen',
      searchIconLabel: 'Buscar',
      searchInputLabel: 'Ingresa lo que quieras encontrar',
      searchPlaceholder: 'Buscar productos, marcas y más…',
      searchPlaceholderMobile: 'Estoy buscando…',
    },
    footer: {
      items: ['tyc', 'politics', 'helpMain'],
    },
    styles: {
      useProximaNova: true,
    },
  },
  MNI: {
    i18n: {
      favorites: 'Favoritos',
      signinAs: 'Ingresaste como',
      notifications: 'Notificaciones',
      noNotifications: 'Nada por acá',
      myAccountDashboard: 'Resumen',
      searchIconLabel: 'Buscar',
      searchInputLabel: 'Ingresa lo que quieras encontrar',
      searchPlaceholder: 'Buscar productos, marcas y más…',
      searchPlaceholderMobile: 'Estoy buscando…',
    },
    footer: {
      items: ['tyc', 'politics', 'helpMain'],
    },
    styles: {
      useProximaNova: true,
    },
  },
  MPA: {
    i18n: {
      favorites: 'Favoritos',
      signinAs: 'Ingresaste como',
      notifications: 'Notificaciones',
      noNotifications: 'Nada por acá',
      myAccountDashboard: 'Resumen',
      searchIconLabel: 'Buscar',
      searchInputLabel: 'Ingresa lo que quieras encontrar',
      searchPlaceholder: 'Buscar productos, marcas y más…',
      searchPlaceholderMobile: 'Estoy buscando…',
    },
    footer: {
      items: ['tyc', 'politics', 'helpMain'],
    },
    styles: {
      useProximaNova: true,
    },
  },
  MPY: {
    i18n: {
      favorites: 'Favoritos',
      signinAs: 'Ingresaste como',
      notifications: 'Notificaciones',
      noNotifications: 'Nada por acá',
      myAccountDashboard: 'Resumen',
      searchIconLabel: 'Buscar',
      searchInputLabel: 'Ingresa lo que quieras encontrar',
      searchPlaceholder: 'Buscar productos, marcas y más…',
      searchPlaceholderMobile: 'Estoy buscando…',
    },
    footer: {
      items: ['tyc', 'politics', 'helpMain'],
    },
    styles: {
      useProximaNova: true,
    },
  },
  MPE: {
    i18n: {
      favorites: 'Favoritos',
      signinAs: 'Ingresaste como',
      notifications: 'Notificaciones',
      noNotifications: 'Nada por acá',
      myAccountDashboard: 'Resumen',
      searchIconLabel: 'Buscar',
      searchInputLabel: 'Ingresa lo que quieras encontrar',
      searchPlaceholder: 'Buscar productos, marcas y más…',
      searchPlaceholderMobile: 'Estoy buscando…',
    },
    footer: {
      items: ['jobs', 'tyc', 'politics', 'helpMain'],
    },
    styles: {
      useProximaNova: true,
    },
  },
  MPT: {
    i18n: {
      favorites: 'Favoritos',
      signinAs: 'Assinado como',
      notifications: 'Avisos',
      noNotifications: ' Por enquanto, não há nada aqui',
      myAccountDashboard: 'Resumen',
      searchIconLabel: 'Buscar',
      searchInputLabel: 'Ingresa lo que quieras encontrar',
      searchPlaceholder: 'Buscar productos, marcas y más…',
      searchPlaceholderMobile: 'Estoy buscando…',
    },
    footer: {
      items: ['tyc', 'politics', 'helpMain'],
    },
    styles: {
      useProximaNova: true,
    },
  },
  MSV: {
    i18n: {
      favorites: 'Favoritos',
      signinAs: 'Ingresaste como',
      notifications: 'Notificaciones',
      noNotifications: 'Nada por acá',
      myAccountDashboard: 'Resumen',
      searchIconLabel: 'Buscar',
      searchInputLabel: 'Ingresa lo que quieras encontrar',
      searchPlaceholder: 'Buscar productos, marcas y más…',
      searchPlaceholderMobile: 'Estoy buscando…',
    },
    footer: {
      items: ['tyc', 'politics', 'helpMain'],
    },
    styles: {
      useProximaNova: true,
    },
  },
  CBT: {
    i18n: {
      signinAs: 'Sing in as',
      myAccountDashboard: 'Dashboard',
    },
    styles: {
      useProximaNova: true,
    },
  },
};
