// frontend-layout/config/platforms/mp.js
module.exports = {
  MLA: {
    i18n: {},
    header: {
      menu: [
        {
          key: 'summary',
          target: '_blank',
        },
        {
          key: 'break',
          isBreak: true,
        },
        {
          key: 'helpMain',
          target: '_blank',
        },
        {
          key: 'logout',
          target: '_self',
        },
      ],
      items: [
        {
          className: 'option-register',
          key: 'registration',
          name: 'Registrate',
          target: '_self',
        },
        {
          className: 'option-help-tablet',
          key: 'helpMain',
          target: '_blank',
        },
        {
          className: 'option-login',
          key: 'login',
          target: '_self',
        },
        {
          className: 'option-register-tablet',
          key: 'registration',
          target: '_self',
        },
        {
          className: 'option-help',
          key: 'helpMain',
          icon: 'nav-icon-help',
          target: '_blank',
        },
      ],
    },
    footerAccess: { hidden: true },
    footer: {
      navigation: ['helpMain', 'tyc', 'developers', 'investor'],
      social: ['facebook', 'twitter', 'linkedin', 'youtube'],
    },
    styles: {
      useProximaNova: true,
    },
  },
  MLB: {
    i18n: {},
    header: {
      menu: [
        {
          key: 'summary',
          target: '_blank',
        },
        {
          key: 'break',
          isBreak: true,
        },
        {
          key: 'helpMain',
          target: '_blank',
        },
        {
          key: 'logout',
          target: '_self',
        },
      ],
      items: [
        {
          className: 'option-register',
          key: 'registration',
          name: 'Cadastre-se',
          target: '_self',
        },
        {
          className: 'option-help-tablet',
          key: 'helpMain',
          target: '_blank',
        },
        {
          className: 'option-login',
          key: 'login',
          target: '_self',
        },
        {
          className: 'option-register-tablet',
          key: 'registration',
          target: '_self',
        },
        {
          className: 'option-help',
          key: 'helpMain',
          icon: 'nav-icon-help',
          target: '_blank',
        },
      ],
    },
    footerAccess: { hidden: true },
    footer: {
      navigation: ['helpMain', 'commercial', 'tyc', 'developers', 'investor'],
      social: ['facebook', 'twitter', 'vimeo', 'youtube', 'linkedin', 'instagram'],
    },
    styles: {
      useProximaNova: true,
    },
  },
  MLM: {
    i18n: {},
    header: {
      menu: [
        {
          key: 'summary',
          target: '_blank',
        },
        {
          key: 'break',
          isBreak: true,
        },
        {
          key: 'helpMain',
          target: '_blank',
        },
        {
          key: 'logout',
          target: '_self',
        },
      ],
      items: [
        {
          className: 'option-register',
          key: 'registration',
          name: 'Registrate',
          target: '_self',
        },
        {
          className: 'option-help-tablet',
          key: 'helpMain',
          target: '_blank',
        },
        {
          className: 'option-login',
          key: 'login',
          target: '_self',
        },
        {
          className: 'option-register-tablet',
          key: 'registration',
          target: '_self',
        },
        {
          className: 'option-help',
          key: 'helpMain',
          icon: 'nav-icon-help',
          target: '_blank',
        },
      ],
    },
    footerAccess: { hidden: true },
    footer: {
      navigation: ['helpMain', 'tyc', 'developers', 'investor', 'politics'],
      social: ['facebook', 'twitter', 'linkedin', 'youtube'],
    },
    styles: {
      useProximaNova: true,
    },
  },
  MLC: {
    i18n: {},
    header: {
      menu: [
        {
          key: 'summary',
          target: '_blank',
        },
        {
          key: 'break',
          isBreak: true,
        },
        {
          key: 'helpMain',
          target: '_blank',
        },
        {
          key: 'logout',
          target: '_self',
        },
      ],
      items: [
        {
          className: 'option-register',
          key: 'registration',
          name: 'Registrate',
          target: '_self',
        },
        {
          className: 'option-help-tablet',
          key: 'helpMain',
          target: '_blank',
        },
        {
          className: 'option-login',
          key: 'login',
          target: '_self',
        },
        {
          className: 'option-register-tablet',
          key: 'registration',
          target: '_self',
        },
        {
          className: 'option-help',
          key: 'helpMain',
          icon: 'nav-icon-help',
          target: '_blank',
        },
      ],
    },
    footerAccess: { hidden: true },
    footer: {
      navigation: ['helpMain', 'tyc', 'developers', 'investor'],
      social: ['facebook', 'twitter', 'linkedin', 'youtube'],
    },
    styles: {
      useProximaNova: true,
    },
  },
  MCO: {
    i18n: {},
    header: {
      menu: [
        {
          key: 'summary',
          target: '_blank',
        },
        {
          key: 'break',
          isBreak: true,
        },
        {
          key: 'helpMain',
          target: '_blank',
        },
        {
          key: 'logout',
          target: '_self',
        },
      ],
      items: [
        {
          className: 'option-register',
          key: 'registration',
          name: 'Registrate',
          target: '_self',
        },
        {
          className: 'option-help-tablet',
          key: 'helpMain',
          target: '_blank',
        },
        {
          className: 'option-login',
          key: 'login',
          target: '_self',
        },
        {
          className: 'option-register-tablet',
          key: 'registration',
          target: '_self',
        },
        {
          className: 'option-help',
          key: 'helpMain',
          icon: 'nav-icon-help',
          target: '_blank',
        },
      ],
    },
    footerAccess: { hidden: true },
    footer: {
      navigation: ['helpMain', 'tyc', 'developers', 'investor'],
      social: ['facebook', 'twitter', 'linkedin', 'youtube'],
    },
    styles: {
      useProximaNova: true,
    },
  },
  MLV: {
    i18n: {},
    header: {
      menu: [
        {
          key: 'summary',
          target: '_blank',
        },
        {
          key: 'break',
          isBreak: true,
        },
        {
          key: 'helpMain',
          target: '_blank',
        },
        {
          key: 'logout',
          target: '_self',
        },
      ],
      items: [
        {
          className: 'option-register',
          key: 'registration',
          name: 'Registrate',
          target: '_self',
        },
        {
          className: 'option-help-tablet',
          key: 'helpMain',
          target: '_blank',
        },
        {
          className: 'option-login',
          key: 'login',
          target: '_self',
        },
        {
          className: 'option-register-tablet',
          key: 'registration',
          target: '_self',
        },
        {
          className: 'option-help',
          key: 'helpMain',
          icon: 'nav-icon-help',
          target: '_blank',
        },
      ],
    },
    footerAccess: { hidden: true },
    footer: {
      navigation: ['helpMain', 'tyc', 'developers', 'investor'],
      social: ['facebook', 'twitter', 'linkedin', 'youtube'],
    },
    styles: {
      useProximaNova: true,
    },
  },
  MLU: {
    i18n: {},
    header: {
      menu: [
        {
          key: 'summary',
          target: '_blank',
        },
        {
          key: 'break',
          isBreak: true,
        },
        {
          key: 'helpMain',
          target: '_blank',
        },
        {
          key: 'logout',
          target: '_self',
        },
      ],
      items: [
        {
          className: 'option-register',
          key: 'registration',
          name: 'Registrate',
          target: '_self',
        },
        {
          className: 'option-help-tablet',
          key: 'helpMain',
          target: '_blank',
        },
        {
          className: 'option-login',
          key: 'login',
          target: '_self',
        },
        {
          className: 'option-register-tablet',
          key: 'registration',
          target: '_self',
        },
        {
          className: 'option-help',
          key: 'helpMain',
          icon: 'nav-icon-help',
          target: '_blank',
        },
      ],
    },
    footerAccess: { hidden: true },
    footer: {
      navigation: ['helpMain', 'tyc', 'developers', 'investor'],
      social: ['facebook', 'twitter', 'linkedin', 'youtube'],
    },
    styles: {
      useProximaNova: true,
    },
  },
  MPE: {
    i18n: {},
    header: {
      menu: [
        {
          key: 'summary',
          target: '_blank',
        },
        {
          key: 'break',
          isBreak: true,
        },
        {
          key: 'helpMain',
          target: '_blank',
        },
        {
          key: 'logout',
          target: '_self',
        },
      ],
      items: [
        {
          className: 'option-register',
          key: 'registration',
          name: 'Registrate',
          target: '_self',
        },
        {
          className: 'option-help-tablet',
          key: 'helpMain',
          target: '_blank',
        },
        {
          className: 'option-login',
          key: 'login',
          target: '_self',
        },
        {
          className: 'option-register-tablet',
          key: 'registration',
          target: '_self',

        },
        {
          className: 'option-help',
          key: 'helpMain',
          icon: 'nav-icon-help',
          target: '_blank',
        },
      ],
    },
    footerAccess: { hidden: true },
    footer: {
      navigation: ['helpMain', 'tyc', 'developers', 'investor'],
      social: ['facebook', 'twitter', 'linkedin', 'youtube'],
    },
    styles: {
      useProximaNova: true,
    },
  },
  MEC: {
    i18n: {},
    header: {
      menu: [
        {
          key: 'summary',
          target: '_blank',
        },
        {
          key: 'break',
          isBreak: true,
        },
        {
          key: 'helpMain',
          target: '_blank',
        },
        {
          key: 'logout',
          target: '_self',
        },
      ],
      items: [
        {
          className: 'option-register',
          key: 'registration',
          name: 'Registrate',
          target: '_self',
        },
        {
          className: 'option-help-tablet',
          key: 'helpMain',
          target: '_blank',
        },
        {
          className: 'option-login',
          key: 'login',
          target: '_self',
        },
        {
          className: 'option-register-tablet',
          key: 'registration',
          target: '_self',
        },
        {
          className: 'option-help',
          key: 'helpMain',
          icon: 'nav-icon-help',
          target: '_blank',
        },
      ],
    },
    footerAccess: { hidden: true },
    footer: {
      navigation: ['helpMain', 'tyc', 'developers', 'investor'],
      social: ['facebook', 'twitter', 'linkedin', 'youtube'],
    },
    styles: {
      useProximaNova: true,
    },
  },
};
