// frontend-layout/config/platforms/ms.js
module.exports = {
  MLA: {
    i18n: {
      search: 'Buscar',
      search_placeholder: 'Buscar...',
      notifications: 'Notificaciones',
      shopping_cart: 'Carrito',
      contact: 'Contacto',
      categories: 'Categorías',
      powered_by: 'Powered by Mercado Shops',
    },
    header: {
      items: {
        full: ['summary', 'operations', 'paymentSolutions', 'configuration', 'helpMain'],
        lite: ['summary', 'helpMain'],
      },
    },
    footerAccess: {
      hidden: true,
    },
    footer: {
      items: ['about', 'developers', 'helpMain'],
    },
    styles: {
      useProximaNova: true,
    },
  },
  MLB: {
    i18n: {
      search: 'Buscar',
      search_placeholder: 'Buscar...',
      notifications: 'Notificações',
      shopping_cart: 'Carrinho',
      contact: 'Contato',
      categories: 'Categorias',
      powered_by: 'Powered by Mercado Shops',
    },
    header: {
      items: {
        full: ['summary', 'operations', 'paymentSolutions', 'configuration', 'helpMain'],
        lite: ['summary', 'helpMain'],
      },
    },
    footerAccess: {
      hidden: true,
    },
    footer: {
      items: ['about', 'developers', 'helpMain'],
    },
    styles: {
      useProximaNova: true,
    },
  },
  MLM: {
    i18n: {
      search: 'Buscar',
      search_placeholder: 'Buscar...',
      notifications: 'Notificaciones',
      shopping_cart: 'Carrito',
      contact: 'Contacto',
      categories: 'Categorías',
      powered_by: 'Powered by Mercado Shops',
    },
    header: {
      items: {
        full: ['summary', 'operations', 'paymentSolutions', 'configuration', 'helpMain'],
        lite: ['summary', 'helpMain'],
      },
    },
    footerAccess: {
      hidden: true,
    },
    footer: {
      items: ['about', 'developers', 'helpMain'],
    },
    styles: {
      useProximaNova: true,
    },
  },
  MLC: {
    i18n: {
      search: 'Buscar',
      search_placeholder: 'Buscar...',
      notifications: 'Notificaciones',
      shopping_cart: 'Carrito',
      contact: 'Contacto',
      categories: 'Categorías',
      powered_by: 'Powered by Mercado Shops',
    },
    header: {
      items: {
        full: ['summary', 'operations', 'paymentSolutions', 'configuration', 'helpMain'],
        lite: ['summary', 'helpMain'],
      },
    },
    footerAccess: {
      hidden: true,
    },
    footer: {
      items: ['about', 'developers', 'helpMain'],
    },
    styles: {
      useProximaNova: true,
    },
  },
  MCO: {
    i18n: {
      search: 'Buscar',
      search_placeholder: 'Buscar...',
      notifications: 'Notificaciones',
      shopping_cart: 'Carrito',
      contact: 'Contacto',
      categories: 'Categorías',
      powered_by: 'Powered by Mercado Shops',
    },
    header: {
      items: {
        full: ['summary', 'operations', 'paymentSolutions', 'configuration', 'helpMain'],
        lite: ['summary', 'helpMain'],
      },
    },
    footerAccess: {
      hidden: true,
    },
    footer: {
      items: ['about', 'developers', 'helpMain'],
    },
    styles: {
      useProximaNova: true,
    },
  },
  MLV: {
    i18n: {
      search: 'Buscar',
      search_placeholder: 'Buscar...',
      notifications: 'Notificaciones',
      shopping_cart: 'Carrito',
      contact: 'Contacto',
      categories: 'Categorías',
      powered_by: 'Powered by Mercado Shops',
    },
    header: {
      items: {
        full: ['summary', 'operations', 'paymentSolutions', 'configuration', 'helpMain'],
        lite: ['summary', 'helpMain'],
      },
    },
    footerAccess: {
      hidden: true,
    },
    footer: {
      items: ['about', 'developers', 'helpMain'],
    },
    styles: {
      useProximaNova: true,
    },
  },
};
