// frontend-layout/config/platforms/pi.js
module.exports = {
  MLC: {
    i18n: {
      favorites: 'Favoritos',
      signinAs: 'Ingresaste como',
      notifications: 'Notificaciones',
      noNotifications: 'Por ahora, no hay nada acá.',
      myAccountDashboard: 'Resumen',
    },
    footerAccess: {
      title: 'Más información',
      items: [
        {
          id: 'about',
          title: 'Acerca de',
          items: ['whoWeAre', 'meli'],
        },
        {
          id: 'help',
          title: 'Ayuda',
          items: ['helpMyAccount', 'helpPublishing', 'helpAdvicesForBuyers'],
        },
        {
          id: 'otherSites',
          title: 'Otros sitios',
          items: ['meliIdeas', 'developers'],
        },
        {
          id: 'socialNetworks',
          title: 'Redes sociales',
          items: ['twitter', 'facebook'],
        },
        {
          id: 'myAccount',
          title: 'Mi cuenta',
        },
      ],
    },
    footer: {
      items: ['workWithUs', 'tyc'],
    },
    styles: {
      useProximaNova: true,
    },
    mobileHeaderMenuOverride: {
      guest: [
        ['home', 'history', 'mapSearch', 'helpMain'],
      ],
      logged: [
        ['home', 'myAccount', 'history', 'myQuotations', 'bookmarks'],
        ['mapSearch'],
        ['helpMain'],
        ['logout'],
      ],
    },
    footerUrls: [
      'myAccount',
      'bookmarks',
      'helpMain',
      'sell',
    ],
  },
};
