// frontend-layout/config/platforms/tl.js
module.exports = {
  MCO: {
    i18n: {
      favorites: 'Favoritos',
      signinAs: 'Ingresaste como',
      notifications: 'Notificaciones',
      noNotifications: 'Nada por acá',
    },
    footerAccess: {
      title: 'Más Información',
      items: [
        {
          id: 'help',
          title: 'Ayuda',
          items: ['helpMain', 'helpContact', 'helpSell'],
        },
        {
          id: 'myAccount',
          title: 'Mi cuenta',
        },
      ],
    },
    footer: {
      items: ['jobs', 'tyc', 'politics', 'helpMain'],
    },
    styles: {
      useProximaNova: false,
    },
  },
  MLV: {
    i18n: {
      favorites: 'Favoritos',
      signinAs: 'Ingresaste como',
      notifications: 'Notificaciones',
      noNotifications: 'Nada por acá',
    },
    footerAccess: {
      title: 'Más Información',
      items: [
        {
          id: 'help',
          title: 'Ayuda',
          items: ['helpMain', 'helpContact', 'helpSell'],
        },
        {
          id: 'myAccount',
          title: 'Mi cuenta',
        },
      ],
    },
    footer: {
      items: ['jobs', 'tyc', 'politics', 'helpMain'],
    },
    styles: {
      useProximaNova: false,
    },
  },
};
