// nordic-attribution/src/AttributionScript.js
/**
 * Module dependencies
 */
/* eslint-disable no-unused-vars */
const React = require('react');
/* eslint-enable no-unused-vars */
const PropTypes = require('prop-types');
const getAttributionScript = require('./get-attribution-script');

/**
 * AttributionScript
 */
function AttributionScript({ browser, platform, cdnDomain, queryString = {}, pageId, referrer }) {
  if (browser && platform && cdnDomain) {
    if (platform.id && platform.id.toLowerCase() === 'ms') {
      return null;
    }

    const script = getAttributionScript(
      browser,
      platform.siteId.toLowerCase(),
      cdnDomain,
      queryString.matt_tool,
      queryString.matt_ignore,
      // Indicates if traffic comes from internal system (pmstrk). internal_referrer='true'
      queryString.internal_referrer,
      pageId,
      // Http referrer
      referrer,
    );
    if (script) {
      return script;
    }
  }
  return null;
}

/**
 * AttributionScript PropsTypes
 */
AttributionScript.PropsTypes = {
  browser: PropTypes.shape({
    name: PropTypes.string.isRequired,
    major: PropTypes.string.isRequired,
  }).isRequired,
  platform: PropTypes.shape({
    id: PropTypes.string,
    siteId: PropTypes.string.isRequired,
  }).isRequired,
  cdnDomain: PropTypes.string.isRequired,
  queryString: PropTypes.shape({
    matt_tool: PropTypes.int,
    matt_ignore: PropTypes.string,
    internal_referrer: PropTypes.string,
  }),
  pageId: PropTypes.string,
  referrer: PropTypes.string,
};

/**
 * Expose AttributionScript
 */
module.exports = AttributionScript;
