// pi-web-navigation/src/navigation/header/components/user-menu/desktop/logged-in/props.js
/* eslint-disable max-len */
const userWidgetProps = require('../../../../../widgets/user-menu/props');
const welcomeProps = require('./welcome/props');
const { mapLinksToUrls } = require('../../../../../utils/links');

// TODO: move to conf
const bookmarks = 'bookmarks';
const sell = 'sell';
const help = 'helpMain';
const userLinks = [
  bookmarks,
  help,
  sell,
];

const loggedInUserMenuProps = (props) => {
  const { navigationUrls: urls, actualLocation } = props;
  const links = mapLinksToUrls({
    links: userLinks,
    urls,
    currentUrl: actualLocation,
    originalProps: props,
    hash: 'nav-header',
    className: (link) => {
      switch (link) {
        case help:
          return 'option-help';
        case sell:
          return 'option-publish';
        default:
          return `option-${link}`;
      }
    },
  });

  return {
    welcome: welcomeProps(props),
    userWidget: userWidgetProps(props),
    userLinks: {
      links,
    },
  };
};

module.exports = loggedInUserMenuProps;
