// mp-web-navigation/src/navigation/header/components/user-menu/guest/props.js
/* eslint-disable security/detect-object-injection */
const { getValue } = require('mp-navigation-config');
const { isHidden } = require('../../../../../layout-type-detection');
const getNavHeaderItems = require('../../../../utils/get-navheader-items');
const mapLinksToUrls = require('../../../../utils/map-links-to-urls');
const navigationProps = require('../navigation/props');
const landingNavigationProps = require('../navigation/landing-navigation/props');

const guestUserMenu = (props) => {
  const { navigationUrls, platform, req, device } = props;
  const navHeader = getValue({
    key: 'navHeader',
    req,
    platform,
  });

  if (isHidden(props.type)) return {};
  if (!navHeader) return {};

  const links = mapLinksToUrls({
    urls: navigationUrls,
    items: getNavHeaderItems(navHeader, device.type, 'guest'),
  });

  return {
    links,
    navigation: navigationProps(props),
    siteId: platform.siteId,
    deviceType: device.type,
    landingNavigation: landingNavigationProps(props),
  };
};

module.exports = guestUserMenu;
